{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Virus-Host Species Relation Extraction\n",
    "## Notebook 2\n",
    "### UC Davis Epicenter for Disease Dynamics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "import os\n",
    "import numpy as np\n",
    "from snorkel import SnorkelSession\n",
    "session = SnorkelSession()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snorkel.models import candidate_subclass\n",
    "\n",
    "VirusHost = candidate_subclass('VirusHost', ['virus', 'host'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part I: Writing Labeling Functions\n",
    "\n",
    "Labeling functions encode our heuristics and weak supervision signals to generate (noisy) labels for our training candidates.\n",
    "\n",
    "In Snorkel, our primary interface through which we provide training signal to the end extraction model we are training is by writing **labeling functions (LFs)** (as opposed to hand-labeling massive training sets). \n",
    "\n",
    "A labeling function is just a Python function that accepts a `Candidate` and returns `1` to mark the `Candidate` as true, `-1` to mark the `Candidate` as false, and `0` to abstain from labeling the `Candidate`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from snorkel.lf_helpers import (\n",
    "    get_left_tokens, \n",
    "    get_right_tokens, \n",
    "    get_between_tokens,\n",
    "    get_text_between, \n",
    "    get_tagged_text,\n",
    "    rule_regex_search_tagged_text,\n",
    "    rule_regex_search_btw_AB,\n",
    "    rule_regex_search_btw_BA,\n",
    "    rule_regex_search_before_A,\n",
    "    rule_regex_search_before_B,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing labels\n",
    "train_cands = session.query(VirusHost).filter(VirusHost.split == 0).order_by(VirusHost.id).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VirusHost(Span(\"b'(TRBV)'\", sentence=8347, chars=[794,799], words=[182,184]), Span(\"b'swine'\", sentence=8347, chars=[969,973], words=[229,229]))\n",
      "Sentence(Document Hubalek-2012-Tick-borne viruses in Europe.pdf,0,b'\"REVIEW Tick-borne viruses in Europe Zdenek Hub\\xc3\\xa1lek & Ivo Rudolf Received: 23 January 2012 /Accepted: 20 March 2012 /Published online: 18 April 2012 # Springer-Verlag 2012 Abstract The aim of this review is to present briefly back- ground information on 27 tick-borne viruses (\\xe2\\x80\\x9ctiboviruses\\xe2\\x80\\x9d) that have been detected in Europe, viz flaviviruses tick- borne encephalitis (TBEV), louping-ill (LIV), Tyuleniy (TYUV), and Meaban (MEAV); orthobunyaviruses Bahig (BAHV) and Matruh (MTRV); phleboviruses Grand Arbaud (GAV), Ponteves (PTVV), Uukuniemi (UUKV), Zaliv Terpeniya (ZTV), and St. Abb\\'s Head (SAHV); nairoviruses Soldado (SOLV), Puffin Island (PIV), Avalon (AVAV), Clo Mor (CMV), Crimean-Congo hemorrhagic fever (CCHFV); bunyavirus Bhanja (BHAV); coltivirus Eyach (EYAV); orbi- viruses Tribec (TRBV), Okhotskiy (OKHV), Cape Wrath (CWV), Mykines (MYKV), Tindholmur (TDMV), and Bau- line (BAUV); two thogotoviruses (Thogoto THOV, Dhori DHOV); and one asfivirus (African swine fever virus ASFV).')\n",
      "Candidate LEFT tokens:   \t ['fever', '(', 'cchfv', ')', ';', 'bunyavirus', 'bhanja', '(', 'bhav', ')', ';', 'coltivirus', 'eyach', '(', 'eyav', ')', ';', 'orbi-', 'viruses', 'tribec']\n",
      "Candidate RIGHT tokens:  \t ['fever', 'virus', 'asfv', ')', '.']\n",
      "Candidate BETWEEN tokens:\t , Okhotskiy (OKHV), Cape Wrath (CWV), Mykines (MYKV), Tindholmur (TDMV), and Bau- line (BAUV); two thogotoviruses (Thogoto THOV, Dhori DHOV); and one asfivirus (African \n"
     ]
    }
   ],
   "source": [
    "# choose a candidate for testing\n",
    "print(train_cands[250])\n",
    "sentence = train_cands[250].get_parent()\n",
    "document = sentence.get_parent()\n",
    "print(sentence)\n",
    "print(\"Candidate LEFT tokens:   \\t\", list(get_left_tokens(train_cands[250],window=20)))\n",
    "print(\"Candidate RIGHT tokens:  \\t\", list(get_right_tokens(train_cands[250],window=20)))\n",
    "print(\"Candidate BETWEEN tokens:\\t\", get_text_between(train_cands[250]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Text Pattern based labeling functions, which look for certain keywords\n",
    "\n",
    "# List to parenthetical\n",
    "def ltp(x):\n",
    "    return '(' + '|'.join(x) + ')'\n",
    "\n",
    "# --------------------------------\n",
    "\n",
    "# Positive LFs:\n",
    "\n",
    "detect = {'detect', 'detects', 'detected', 'detecting', 'detection', 'detectable'}\n",
    "detect_l = ['detect', 'detects', 'detected', 'detecting', 'detection', 'detectable']\n",
    "infect = {'infect', 'infects', 'infected', 'infecting', 'infection'}\n",
    "isolate = {'isolate', 'isolates', 'isolated', 'isolating', 'isolation'}\n",
    "other_verbs = {\n",
    "    'transmit(ted)?', 'found', 'find', 'findings', 'remove(d)?', 'affect(s|ed|ing)?', 'confirm(s|ed|ing)?', 'relat(ed|es|e|ing|ion)?', 'recovered', 'identified',\n",
    "}\n",
    "misc = {\n",
    "    'seropositive', 'seropositivity', 'positive', 'host(s)?', 'prevalen(ce|t)?', 'case(s)?', 'ELISA', 'titer', 'viremia', 'antibod(y|ies)?', 'antigen', 'exposure', 'PCR', 'polymerase chain reaction', 'RNA', 'DNA', 'nucleotide', 'sequence', 'evidence', 'common', 'success', 'successfully', 'extract', 'PFU', '(PFU)', 'plaque-forming unit'\n",
    "}\n",
    "\n",
    "causal = ['caus(es|ed|e|ing|ation)?', 'induc(es|ed|e|ing)?', 'associat(ed|ing|es|e|ion)?']\n",
    "\n",
    "positive = {'detect', 'detects', 'detected', 'detecting', 'detection', 'detectable', 'infect', 'infects', 'infected', 'infecting', 'infection', 'isolate', 'isolates', 'isolated', 'isolating', 'isolation'}\n",
    "positive_l = ['detect', 'detects', 'detected', 'detecting', 'detection', 'detectable', 'infect', 'infects', 'infected', 'infecting', 'infection', 'isolate', 'isolates', 'isolated', 'isolating', 'isolation']\n",
    "negative = {\n",
    "    'negative (antibodies)?', 'negate', '\\W+not\\W+', '\\W+no\\W+', '(titer\\W+(?:\\w+\\W+){1,6}?less than)', 'none', 'resist'\n",
    "}\n",
    "negative_l = [\n",
    "    'negative (antibodies)?', 'negate', '\\W+not\\W+', '\\W+no\\W+', '(titer\\W+(?:\\w+\\W+){1,6}?less than)', 'none', 'resist'\n",
    "]\n",
    "neg_rgx = r'|'.join(negative_l)\n",
    "\n",
    "\n",
    "\n",
    "# words like detect \n",
    "def LF_detect(c):\n",
    "    if (len(detect.intersection(get_between_tokens(c))) > 0) and (len(negative.intersection(get_between_tokens(c))) == 0): \n",
    "        return 1\n",
    "    elif (len(detect.intersection(get_left_tokens(c[0], window=20))) > 0) and (len(negative.intersection(get_left_tokens(c[0], window=20))) == 0):\n",
    "        return 1\n",
    "    elif (len(detect.intersection(get_left_tokens(c[1], window=20))) > 0) and (len(negative.intersection(get_left_tokens(c[1], window=20))) == 0):\n",
    "        return 1\n",
    "    elif (len(detect.intersection(get_right_tokens(c[0], window=20))) > 0) and ((len(negative.intersection(get_right_tokens(c[0], window=20))) == 0)):\n",
    "        return 1\n",
    "    elif (len(detect.intersection(get_right_tokens(c[1], window=20))) > 0) and ((len(negative.intersection(get_right_tokens(c[1], window=20))) == 0)):\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def LF_detect2(c):\n",
    "    return 1 if (\n",
    "        re.search(r'{{A}}.{0,50}' + ltp(detect_l) + '.{0,50}{{B}}', get_tagged_text(c), re.I) \n",
    "        and not re.search('{{A}}.{0,50}(not|no).{0,20}' + ltp(detect_l) + '.{0,50}{{B}}', get_tagged_text(c), re.I)\n",
    "    ) else 0\n",
    "\n",
    "    \n",
    "def LF_infect(c):\n",
    "    if len(infect.intersection(get_between_tokens(c))) > 0: \n",
    "        return 1\n",
    "    elif len(infect.intersection(get_left_tokens(c[0], window=20))) > 0:\n",
    "        return 1\n",
    "    elif len(infect.intersection(get_left_tokens(c[1], window=20))) > 0:\n",
    "        return 1\n",
    "    elif len(infect.intersection(get_right_tokens(c[0], window=20))) > 0:\n",
    "        return 1\n",
    "    elif len(infect.intersection(get_right_tokens(c[1], window=20))) > 0:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "    # Words like 'isolated'\n",
    "def LF_isolate(c):\n",
    "    if len(isolate.intersection(get_between_tokens(c))) > 0: \n",
    "        return 1\n",
    "    elif len(isolate.intersection(get_left_tokens(c[0], window=20))) > 0:\n",
    "        return 1\n",
    "    elif len(isolate.intersection(get_left_tokens(c[1], window=20))) > 0:\n",
    "        return 1\n",
    "    elif len(isolate.intersection(get_right_tokens(c[0], window=20))) > 0:\n",
    "        return 1\n",
    "    elif len(isolate.intersection(get_right_tokens(c[1], window=20))) > 0:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "        \n",
    "def LF_misc(c):\n",
    "    if (len(misc.intersection(get_between_tokens(c))) > 0) and not (re.search(neg_rgx, get_tagged_text(c), re.I)): \n",
    "        return 1\n",
    "    elif (len(misc.intersection(get_left_tokens(c[0], window=20))) > 0) and not (re.search(neg_rgx, get_tagged_text(c), re.I)): \n",
    "        return 1\n",
    "    elif (len(misc.intersection(get_left_tokens(c[1], window=20))) > 0) and not (re.search(neg_rgx, get_tagged_text(c), re.I)): \n",
    "        return 1\n",
    "    elif (len(misc.intersection(get_right_tokens(c[0], window=20))) > 0) and not (re.search(neg_rgx, get_tagged_text(c), re.I)): \n",
    "        return 1\n",
    "    elif (len(misc.intersection(get_right_tokens(c[1], window=20))) > 0) and not (re.search(neg_rgx, get_tagged_text(c), re.I)): \n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "# Words like 'caused'\n",
    "def LF_v_cause_h(c):\n",
    "    return 1 if (\n",
    "        re.search(r'{{A}}.{0,50} ' + ltp(causal) + '.{0,50}{{B}}', get_tagged_text(c), re.I)\n",
    "        and not re.search('{{A}}.{0,50}(not|no|negative).{0,20}' + ltp(causal) + '.{0,50}{{B}}', get_tagged_text(c), re.I)\n",
    "    ) else 0\n",
    "\n",
    "# if candidates are nearby and check for negative words\n",
    "def LF_v_h(c):\n",
    "    return 1 if (\n",
    "        re.search(r'{{A}}.{0,250}{{B}}', get_tagged_text(c), re.I)\n",
    "        and not re.search(neg_rgx, get_tagged_text(c), re.I)\n",
    "    ) else 0\n",
    "\n",
    "def LF_h_v(c):\n",
    "    return 1 if (\n",
    "        re.search(r'{{B}}.{0,250}{{A}}', get_tagged_text(c), re.I)\n",
    "        and not re.search(neg_rgx, get_tagged_text(c), re.I)\n",
    "    ) else 0\n",
    "\n",
    "# positive verbs (detect, infect, isolate)\n",
    "def LF_positive(c):\n",
    "    if (len(positive.intersection(get_between_tokens(c))) > 0) and (len(negative.intersection(get_between_tokens(c))) == 0): \n",
    "        return 1\n",
    "    elif (len(positive.intersection(get_left_tokens(c[0], window=20))) > 0)  and (len(negative.intersection(get_left_tokens(c[0], window=15))) == 0):\n",
    "        return 1\n",
    "    elif (len(positive.intersection(get_left_tokens(c[1], window=20))) > 0) and (len(negative.intersection(get_left_tokens(c[1], window=20))) == 0):\n",
    "        return 1\n",
    "    elif (len(positive.intersection(get_right_tokens(c[0], window=20))) > 0) and ((len(negative.intersection(get_right_tokens(c[0], window=20))) == 0)):\n",
    "        return 1\n",
    "    elif (len(positive.intersection(get_right_tokens(c[1], window=20))) > 0) and ((len(negative.intersection(get_right_tokens(c[1], window=20))) == 0)):\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def LF_positive2(c):\n",
    "    return 1 if (\n",
    "        re.search(r'{{A}}.{0,100} ' + ltp(positive_l) + '.{0,100}{{B}}', get_tagged_text(c), re.I)\n",
    "        and not re.search('{{A}}.{0,100}(not|no|negative).{0,20}' + ltp(positive_l) + '.{0,100}{{B}}', get_tagged_text(c), re.I)\n",
    "    ) else 0\n",
    "\n",
    "def LF_other_verbs(c):\n",
    "    if (len(other_verbs.intersection(get_between_tokens(c))) > 0) and (len(negative.intersection(get_between_tokens(c))) == 0): \n",
    "        return 1\n",
    "    elif (len(other_verbs.intersection(get_left_tokens(c[0], window=20))) > 0)  and (len(negative.intersection(get_left_tokens(c[0], window=20))) == 0):\n",
    "        return 1\n",
    "    elif (len(other_verbs.intersection(get_left_tokens(c[1], window=20))) > 0) and (len(negative.intersection(get_left_tokens(c[1], window=20))) == 0):\n",
    "        return 1\n",
    "    elif (len(other_verbs.intersection(get_right_tokens(c[0], window=20))) > 0) and ((len(negative.intersection(get_right_tokens(c[0], window=20))) == 0)):\n",
    "        return 1\n",
    "    elif (len(other_verbs.intersection(get_right_tokens(c[1], window=20))) > 0) and ((len(negative.intersection(get_right_tokens(c[1], window=20))) == 0)):\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "    \n",
    "def LF_percents(c):\n",
    "    return 1 if (\n",
    "        re.search(r'{{A}}.{0,100}' + '\\d+%|\\d+(.)?\\d+%' + '.{0,100}{{B}}', get_tagged_text(c), re.I) \n",
    "        and not re.search('(none|not|\\W+no\\W+|humidity)', get_text_between(c), re.I)\n",
    "    ) else 0\n",
    "\n",
    "\n",
    "# -----------------------------------\n",
    "\n",
    "# Negative LFs:\n",
    "\n",
    "# Uncertain pairs\n",
    "uncertain = ['combin', 'possible', 'unlikely']\n",
    "\n",
    "def LF_uncertain(c):\n",
    "    return rule_regex_search_before_A(c, ltp(uncertain) + '.*', -1)\n",
    "\n",
    "# if candidate pair is too far apart (between 200-5000 characs apart), mark as negative\n",
    "def LF_far_v_h(c):\n",
    "    return rule_regex_search_btw_AB(c, '.{250,5000}', -1)\n",
    "\n",
    "def LF_far_h_v(c):\n",
    "    return rule_regex_search_btw_BA(c, '.{250,5000}', -1)\n",
    "\n",
    "def LF_neg_h(c):\n",
    "    return -1 if re.search('(none|not|no) .{0,100}{{B}}', get_tagged_text(c), flags=re.I) else 0\n",
    "\n",
    "def LF_neg_assertions(c):\n",
    "    if (len(negative.intersection(get_between_tokens(c))) > 0): \n",
    "        return -1\n",
    "    elif (len(negative.intersection(get_left_tokens(c[0], window=20))) > 0):\n",
    "        return -1\n",
    "    elif (len(negative.intersection(get_left_tokens(c[1], window=20))) > 0):\n",
    "        return -1\n",
    "    elif (len(negative.intersection(get_right_tokens(c[0], window=20))) > 0):\n",
    "        return -1\n",
    "    elif (len(negative.intersection(get_right_tokens(c[1], window=20))) > 0):\n",
    "        return -1\n",
    "    else:\n",
    "        return 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distant Supervision LFs\n",
    "# Compare candidates with a database of known virus-host pairs (from Virus-Host Database)\n",
    "\n",
    "import bz2\n",
    "\n",
    "# Function to remove special characters from text\n",
    "def strip_special(s):\n",
    "    return ''.join(c for c in s if ord(c) < 128)\n",
    "\n",
    "# Read in known pairs and save as set of tuples\n",
    "with bz2.BZ2File('virushostdb.tar.bz2', 'rb') as f:\n",
    "    known_pairs = set(\n",
    "        tuple(strip_special(x.decode('utf-8')).strip().split('\\t')) for x in f.readlines()\n",
    "    )\n",
    "\n",
    "def LF_distant_supervision(c):\n",
    "    v, h = c.virus.get_span(), c.host.get_span()\n",
    "    return 1 if (v,h) in known_pairs else 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('Salmonella phage vB_SnwM_CGG4-1',\n",
       "  'Salmonella enterica subsp. enterica serovar Newport'),\n",
       " ('Pseudomonas phage vB_Pae_PS44', 'Pseudomonas aeruginosa'),\n",
       " ('Peanut mottle virus', 'Pisum sativum'),\n",
       " ('Leonurus mosaic virus', 'Lamiaceae'),\n",
       " ('East African cassava mosaic Malawi virus', 'Manihot esculenta'),\n",
       " ('Mycobacterium phage Validus', 'Mycolicibacterium smegmatis'),\n",
       " ('Zostera marina amalgavirus 2', 'Zostera marina'),\n",
       " ('Bush viper reovirus', 'Atheris squamigera'),\n",
       " ('Chaetoceros tenuissimus RNA virus type-II', 'Chaetoceros tenuissimus'),\n",
       " ('Burkholderia phage KS10', 'Burkholderia cenocepacia'),\n",
       " ('New World begomovirus associated satellite DNA isolate 404N1',\n",
       "  'Sidastrum micranthum'),\n",
       " ('Mycobacterium virus Eureka', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Wenzhou hepe-like virus 2', 'root'),\n",
       " ('Acidovorax phage ACP17', 'Acidovorax citrulli'),\n",
       " ('Stenotrophomonas phage PSH1', 'Stenotrophomonas'),\n",
       " ('Escherichia phage PBECO 4', 'Escherichia coli O157'),\n",
       " ('Carnation mottle virus', 'Dianthus caryophyllus'),\n",
       " ('Human adenovirus 1', 'Homo sapiens'),\n",
       " ('Hubei Poty-like virus 1', 'Diptera'),\n",
       " ('Bivalve RNA virus G4', 'Bivalvia'),\n",
       " ('Lettuce infectious yellows virus', 'Nicotiana clevelandii'),\n",
       " ('Ugandan cassava brown streak virus', 'Nicotiana benthamiana'),\n",
       " ('Cherry green ring mottle virus', 'Prunus serrulata var. lannesiana'),\n",
       " ('Norovirus Hu/Chiba/04-1050/2005/JP', 'Homo sapiens'),\n",
       " ('Rosellinia necatrix partitivirus 6', 'Rosellinia necatrix'),\n",
       " ('Shewanella sp. phage 1/44', 'Shewanella sp.'),\n",
       " ('Bunyamwera virus', 'Homo sapiens'),\n",
       " ('Chaetoceros socialis forma radians RNA virus 1', 'Chaetoceros socialis'),\n",
       " ('Koala retrovirus', 'Phascolarctos cinereus'),\n",
       " ('Norovirus Hu/Norwalk/10285/2010/VNM', 'Homo sapiens'),\n",
       " ('Curionopolis virus', 'Culicoides'),\n",
       " ('Maize chlorotic dwarf virus', 'Zea mays'),\n",
       " ('Siegesbeckia yellow vein virus-[GD13]-associated DNA beta',\n",
       "  'Sigesbeckia glabrescens'),\n",
       " ('Human herpesvirus 4 type 2', 'Homo sapiens'),\n",
       " ('Brucella phage R/C', 'Brucella canis'),\n",
       " ('Campylobacter phage vB_CjeM_Los1', 'Campylobacter jejuni'),\n",
       " ('Choristoneura murinana nucleopolyhedrovirus', 'Choristoneura murinana'),\n",
       " ('Escherichia phage EK99P-1', 'Escherichia coli'),\n",
       " ('Helenium virus S', 'Helenium amarum'),\n",
       " ('Hubei sobemo-like virus 45', 'root'),\n",
       " ('Human papillomavirus type 163', 'Homo sapiens'),\n",
       " ('Pseudoalteromonas phage HM1', 'Pseudoalteromonas sp.'),\n",
       " ('Kyuri green mottle mosaic virus', 'Cucurbita maxima'),\n",
       " ('Tipula oleracea nudivirus', 'Tipula oleracea'),\n",
       " ('Lake Sarah-associated circular virus-46', 'root'),\n",
       " ('Lettuce mosaic virus', 'Pisum sativum'),\n",
       " ('Burkholderia virus Bcep781', 'Burkholderia cepacia'),\n",
       " ('Maize streak virus - E[Pat]', 'Eleusine indica'),\n",
       " ('Grapevine yellow speckle viroid 2', 'Vitis vinifera'),\n",
       " ('Orchid fleck dichorhavirus', 'Coelogyne'),\n",
       " ('Hepatitis C virus subtype 4v', 'Homo sapiens'),\n",
       " ('Nilaparvata lugens honeydew virus-2', 'Nilaparvata lugens'),\n",
       " ('Bovine astrovirus CH13', 'Bos taurus'),\n",
       " ('Kunsagivirus B', 'Eidolon helvum'),\n",
       " ('Spissistilus festinus virus 1', 'Spissistilus festinus'),\n",
       " ('Yambean mosaic virus', 'Dioscorea alata'),\n",
       " ('Fig badnavirus 1', 'Ficus'),\n",
       " ('Choristoneura fumiferana DEF multiple nucleopolyhedrovirus', 'Lepidoptera'),\n",
       " ('Human papillomavirus type 16', 'Homo sapiens'),\n",
       " ('Mycobacterium phage Saal', 'Mycolicibacterium smegmatis'),\n",
       " ('Mycobacterium phage Tweety', 'Mycolicibacterium smegmatis'),\n",
       " ('Staphylococcus virus 80alpha', 'Staphylococcus aureus'),\n",
       " ('Carrot mottle virus', 'Nicotiana benthamiana'),\n",
       " ('Cabassou virus', 'Cabassous'),\n",
       " ('Borna disease virus 1', 'Equidae'),\n",
       " ('Norovirus Hu/GII-4/Kumamoto1/2006/JP', 'Homo sapiens'),\n",
       " ('Salmonella virus FSLSP030',\n",
       "  'Salmonella enterica subsp. enterica serovar Dublin'),\n",
       " ('Gordonia phage GMA3', 'Gordonia malaquae'),\n",
       " ('Porcine polyomavirus',),\n",
       " ('Melon necrotic spot virus', 'Cucumis sativus'),\n",
       " ('Pudu puda papillomavirus 1', 'Pudu puda'),\n",
       " ('Culex originated Tymoviridae-like virus', 'Culicidae'),\n",
       " ('Beihai narna-like virus 14', 'root'),\n",
       " ('Escherichia phage pro147', 'Escherichia coli'),\n",
       " ('Norovirus Hu/GII-4/Aichi3/2007/JP', 'Homo sapiens'),\n",
       " ('Bluetongue virus', 'Odocoileus virginianus'),\n",
       " ('Simian adenovirus 6', 'Macaca mulatta'),\n",
       " ('Zaire ebolavirus', 'Epomops franqueti'),\n",
       " ('Rabies lyssavirus', 'Felis catus'),\n",
       " ('Yichang Insect virus', 'Aulacorthum magnoliae'),\n",
       " ('Finkel-Biskis-Jinkins murine sarcoma virus', 'Mus musculus'),\n",
       " ('Phnom Penh bat virus', 'Cynopterus brachyotis'),\n",
       " ('Canine respiratory coronavirus', 'Canis lupus familiaris'),\n",
       " ('Termite associated circular virus 1', 'Termitoidae'),\n",
       " ('Skunkpox virus', 'Mephitis mephitis'),\n",
       " ('Epizootic hemorrhagic disease virus (serotype 2 / strain Alberta)',\n",
       "  'Antilocapra americana'),\n",
       " ('Pectobacterium phage Peat1', 'Pectobacterium atrosepticum'),\n",
       " ('Fowl aviadenovirus A', 'Galliformes'),\n",
       " ('Pseudomonas phage LKD16', 'Pseudomonas aeruginosa'),\n",
       " ('Hubei zhaovirus-like virus 2', 'root'),\n",
       " ('Pseudomonas virus DMS3', 'Pseudomonas aeruginosa'),\n",
       " ('Beihai sobemo-like virus 6', 'root'),\n",
       " ('Ectromelia virus', 'Mus musculus'),\n",
       " ('Wenzhou hepe-like virus 1', 'root'),\n",
       " ('Mimosa yellow leaf curl virus', 'Fabaceae'),\n",
       " ('Feline coronavirus UU30', 'Felis'),\n",
       " ('Aedes alboannulatus toti-like virus 1', 'Ochlerotatus camptorhynchus'),\n",
       " ('Ageratum yellow vein Singapore alphasatellite', 'Ageratum conyzoides'),\n",
       " ('Hubei diptera virus 11', 'root'),\n",
       " ('Rodent arterivirus', 'Neodon clarkei'),\n",
       " ('Turkey avisivirus', 'Meleagris gallopavo'),\n",
       " ('Alternaria brassicicola fusarivirus 1', 'Alternaria brassicicola'),\n",
       " ('Badu phasivirus', 'Culex'),\n",
       " ('Mesta yellow vein mosaic virus-associated DNA beta', 'Hibiscus sabdariffa'),\n",
       " ('Clostridium phage phiCP26F', 'Clostridium perfringens'),\n",
       " ('Sida mosaic Sinaloa virus', 'Sida'),\n",
       " ('Cellulophaga phage phi13:2', 'Cellulophaga baltica'),\n",
       " ('Desmodium yellow mottle virus', 'Desmodium'),\n",
       " ('Lactobacillus phage Lv-1', 'Lactobacillus'),\n",
       " ('Mycobacterium virus Drago', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Simian adenovirus 46', 'Gorilla'),\n",
       " ('Circovirus-like genome DCCV-9', 'root'),\n",
       " ('Sweet potato chlorotic stunt virus', 'Ipomoea batatas'),\n",
       " ('Tobacco curly shoot virus - [Y35]', 'Nicotiana tabacum'),\n",
       " ('Telfairia golden mosaic virus', 'Telfairia occidentalis'),\n",
       " ('Sudan ebolavirus - Nakisamata', 'Homo sapiens'),\n",
       " ('Pineapple mealybug wilt-associated virus 2', 'Ananas comosus'),\n",
       " ('Raspberry ringspot virus', 'Vitis vinifera'),\n",
       " ('Acinetobacter phage AB1', 'Acinetobacter baumannii'),\n",
       " ('Tomato black ring virus', 'Phaseolus vulgaris'),\n",
       " ('Broad bean mottle virus', 'Vicia faba'),\n",
       " ('Mycobacterium phage MOOREtheMARYer', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Enterobacteria phage GEC-3S', 'Escherichia coli O104:H4'),\n",
       " ('Carnation yellow fleck virus', 'Dianthus caryophyllus'),\n",
       " ('Norovirus Hu/Norwalk/10179/2009/VNM', 'Homo sapiens'),\n",
       " ('TT virus sle1957', 'Thyronectria obscura'),\n",
       " ('Beihai sphaeromadae virus 3', 'root'),\n",
       " ('Hubei virga-like virus 9', 'root'),\n",
       " ('Mycobacterium phage Shipwreck', 'Mycolicibacterium smegmatis'),\n",
       " ('Bean golden yellow mosaic virus-[Mexico]', 'Phaseolus vulgaris'),\n",
       " ('Hubei tombus-like virus 30', 'root'),\n",
       " ('Melon yellow spot virus', 'Cucumis melo'),\n",
       " ('Papaya lethal yellowing virus', 'Carica papaya'),\n",
       " ('Borna disease virus 1', 'Hexaprotodon liberiensis'),\n",
       " ('Shahe narna-like virus 3', 'root'),\n",
       " ('Bacillus phage vB_BanS-Tsamsa', 'Bacillus anthracis'),\n",
       " ('Feline coronavirus UU3', 'Felis catus'),\n",
       " ('Hermit crab associated circular virus', 'Anomura'),\n",
       " ('Odonata-associated circular virus 5', 'Libellula quadrimaculata'),\n",
       " ('Cardiospermum yellow leaf curl betasatellite', 'Cardiospermum'),\n",
       " ('Staphylococcus virus 42e', 'Staphylococcus aureus'),\n",
       " ('Camel associated porprismacovirus 3', 'Camelus dromedarius'),\n",
       " ('Acinetobacter phage SH-Ab 15519', 'Acinetobacter baumannii'),\n",
       " ('Lassa mammarenavirus', 'Homo sapiens'),\n",
       " ('Mycobacterium virus Kugel', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Akabane virus', 'Bos taurus'),\n",
       " ('Coconut foliar decay alphasatellite 6', 'Cocos nucifera'),\n",
       " ('Human papillomavirus type 44', 'Homo sapiens'),\n",
       " ('Eubenangee virus', 'Bos taurus'),\n",
       " ('Lettuce infectious yellows virus', 'Cucurbita moschata'),\n",
       " ('Calicivirus pig/AB90/CAN', 'Sus scrofa'),\n",
       " ('Wenling picorna-like virus 2', 'root'),\n",
       " ('Beihai picorna-like virus 20', 'root'),\n",
       " ('Burdock mottle virus', 'Chenopodium quinoa'),\n",
       " ('Flavobacterium sp. phage 1/32', 'Flavobacterium sp.'),\n",
       " ('Guama virus', 'Sapajus apella'),\n",
       " ('Odonata-associated circular virus-17', 'Libellula quadrimaculata'),\n",
       " ('Human parainfluenza virus 4a', 'Homo sapiens'),\n",
       " ('Mopeia Lassa virus reassortant 29', 'Chlorocebus aethiops'),\n",
       " ('Okra yellow vein disease associated sequence', 'Abelmoschus esculentus'),\n",
       " ('Enterobacteria phage JS10', 'Escherichia coli K-12'),\n",
       " ('Avian avulavirus 7', 'Cricetinae'),\n",
       " ('Satellite tobacco necrosis virus 2', 'Phaseolus vulgaris'),\n",
       " ('Mycobacterium phage Chah', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Santa barbara virus', 'Psychodidae'),\n",
       " ('Tomato yellow leaf curl virus', 'Solanum lycopersicum'),\n",
       " ('Clostridium phage phiCT19406B', 'Clostridium'),\n",
       " ('Klebsiella phage Kpn112', 'Klebsiella pneumoniae'),\n",
       " ('Cowpox virus', 'Elephas maximus'),\n",
       " ('Ageratum yellow vein virus-Ishigaki', 'Ageratum houstonianum'),\n",
       " ('Pepper yellow dwarf virus - Mexico', 'Capsicum annuum var. annuum'),\n",
       " ('Brome mosaic virus', 'Bromus inermis'),\n",
       " ('Pseudomonas phage YMC11/06/C171_PPU_BP', 'Pseudomonas putida'),\n",
       " ('Salmonella phage 7-11',\n",
       "  'Salmonella enterica subsp. enterica serovar Newport'),\n",
       " ('Hubei sobemo-like virus 1', 'root'),\n",
       " ('Tamus red mosaic virus', 'Chenopodium quinoa'),\n",
       " ('Bovine viral diarrhea virus type 1b', 'Bos taurus'),\n",
       " ('Orchid fleck dichorhavirus', 'Odontoglossum'),\n",
       " ('Lactate dehydrogenase-elevating virus', 'Mus musculus domesticus'),\n",
       " ('Synechococcus phage syn9', 'Synechococcus'),\n",
       " ('Rhinolophus sinicus bocaparvovirus', 'Rhinolophus sinicus'),\n",
       " ('Norovirus Hu/GII-4/Kumamoto2/2006/JP', 'Homo sapiens'),\n",
       " ('Chlorocebus pygerythrus polyomavirus 3', 'Chlorocebus pygerythrus'),\n",
       " ('Bat coronavirus HKU5-1', 'Chiroptera'),\n",
       " ('Bat associated cyclovirus 4', 'Carollia perspicillata'),\n",
       " ('Groundnut rosette virus', 'Arachis hypogaea'),\n",
       " ('Lactococcus phage P092', 'Lactococcus lactis'),\n",
       " ('Norovirus Hu/Norwalk/20122/2009/VNM', 'Homo sapiens'),\n",
       " ('Streptomyces phage R4', 'Streptomyces albus'),\n",
       " ('Citrobacter phage SH2', 'Citrobacter freundii'),\n",
       " ('Changjiang tombus-like virus 8', 'root'),\n",
       " ('Asystasia mosaic Madagascar virus', 'Asystasia gangetica'),\n",
       " ('Mycobacterium phage Harvey', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Torque teno midi virus 7', 'Homo sapiens'),\n",
       " ('Dabieshan virus', 'Niviventer confucianus'),\n",
       " ('Cherry small circular viroid-like RNA 2', 'Prunus avium'),\n",
       " ('Peanut stripe virus', 'Nicotiana benthamiana'),\n",
       " ('Fusarium graminearum hypovirus 2', 'Fusarium graminearum'),\n",
       " ('Stenotrophomonas phage IME13', 'Stenotrophomonas maltophilia'),\n",
       " ('Streptomyces phage Mildred21', 'Streptomyces viridochromogenes'),\n",
       " ('Circoviridae 8 LDMD-2013', 'root'),\n",
       " ('Kimberley virus', 'Bos taurus'),\n",
       " ('Wenzhou picorna-like virus 45', 'root'),\n",
       " ('Wenzhou qinvirus-like virus 1', 'root'),\n",
       " ('Rhodobacter phage RC1', 'Rhodobacter'),\n",
       " ('Hubei tombus-like virus 34', 'root'),\n",
       " ('Streptococcus phage APCM01', 'Streptococcus'),\n",
       " ('Chocolate lily virus A', 'Dichopogon'),\n",
       " ('Escherichia phage ST0', 'Escherichia coli H8'),\n",
       " ('Escherichia phage WA45', 'Escherichia'),\n",
       " ('Hepatitis C virus subtype 2c', 'Homo sapiens'),\n",
       " ('Beihai picorna-like virus 57', 'root'),\n",
       " ('Lake Sarah-associated circular virus-44', 'root'),\n",
       " ('Rhodobacter phage RcTitan', 'Rhodobacter capsulatus YW1'),\n",
       " ('Pseudomonas virus D3112', 'Pseudomonas aeruginosa'),\n",
       " ('Beet black scorch virus', 'Physalis pubescens'),\n",
       " ('Murine norovirus GV/CR5/2005/USA', 'Mus musculus'),\n",
       " ('Staphylococcus phage SA5', 'Staphylococcus aureus'),\n",
       " ('Streptococcus phage T12', 'Streptococcus pyogenes'),\n",
       " ('Usutu virus', 'Turdus merula'),\n",
       " ('Pseudomonas phage vB_PaeM_E215', 'Pseudomonas'),\n",
       " ('Grapevine viroid', 'Humulus lupulus'),\n",
       " ('Rose spring dwarf-associated virus', 'Liliopsida'),\n",
       " ('Cyanophage KBS-S-2A', 'Synechococcus sp.'),\n",
       " ('Simian pegivirus', 'Piliocolobus tephrosceles'),\n",
       " ('Tomato crinkle yellow leaf virus', 'Solanum lycopersicum'),\n",
       " ('Blackberry chlorotic ringspot virus', 'Fragaria'),\n",
       " ('Lucky bamboo bacilliform virus', 'Dracaena sanderiana'),\n",
       " ('Beihai picobirna-like virus 7', 'root'),\n",
       " ('His 1 virus', 'Haloarcula hispanica'),\n",
       " ('Mycobacterium phage Dreamboat', 'Mycolicibacterium smegmatis'),\n",
       " ('Rasavirus sp.', 'Rattus'),\n",
       " ('Bovine herpesvirus type 1.2 strain B589', 'Bos taurus'),\n",
       " ('Arthrobacter phage Beans', 'Arthrobacter sp. ATCC 21022'),\n",
       " ('Anthoxanthum odoratum amalgavirus 1', 'Anthoxanthum odoratum'),\n",
       " ('Rabbit fibroma virus', 'Oryctolagus cuniculus'),\n",
       " ('Simian endogenous retrovirus', 'Papio cynocephalus'),\n",
       " ('Enterobacteria phage FL68 Tallahassee/FL/2012', 'Escherichia coli'),\n",
       " ('Bat adenovirus TJM', 'Myotis ricketti'),\n",
       " ('Enterococcus phage EFRM31', 'Enterococcus faecalis'),\n",
       " ('Rodent Torque teno virus 3', 'Bolomys lasiurus'),\n",
       " ('Chronic bee paralysis virus', 'Apis mellifera'),\n",
       " ('Seoul orthohantavirus', 'Homo sapiens'),\n",
       " ('Equus caballus papillomavirus 3', 'Equus caballus'),\n",
       " ('Hubei virga-like virus 2', 'root'),\n",
       " ('Streptococcus virus 2972', 'Streptococcus thermophilus'),\n",
       " ('Carnation mottle virus', 'Dianthus chinensis'),\n",
       " ('Propionibacterium phage PHL082M00', 'Propionibacterium'),\n",
       " ('Grapevine virus I', 'Vitis vinifera'),\n",
       " ('Lolium latent virus', 'Lolium multiflorum x Lolium perenne'),\n",
       " ('Cyclovirus ZM01', 'Crocidura hirta'),\n",
       " ('Mycobacterium phage ShedlockHolmes', 'Mycolicibacterium smegmatis'),\n",
       " ('Red clover powdery mildew-associated totivirus 1', 'Erysiphe trifoliorum'),\n",
       " ('Circovirus-like genome DCCV-13', 'root'),\n",
       " ('Rhesus macaque simian foamy virus', 'Macaca mulatta'),\n",
       " ('Foot-and-mouth disease virus - type SAT 2', 'Sus scrofa'),\n",
       " ('Staphylococcus virus 53', 'Staphylococcus aureus'),\n",
       " ('Bombyx mori Macula-like virus', 'Bombyx mori'),\n",
       " ('Urochloa streak virus', 'Brachiaria deflexa'),\n",
       " ('Sindbis-like virus', 'Mus musculus'),\n",
       " ('Mycobacterium phage Bipper', 'Mycolicibacterium smegmatis'),\n",
       " ('Beet virus Q', 'Beta vulgaris'),\n",
       " ('Petunia vein clearing virus', 'Petunia'),\n",
       " ('Wenzhou picorna-like virus 48', 'root'),\n",
       " ('Hubei picorna-like virus 13', 'root'),\n",
       " ('Wenling tombus-like virus 3', 'root'),\n",
       " ('Murine orthopneumovirus', 'Mesocricetus auratus'),\n",
       " ('Strawberry mottle virus', 'Chenopodium quinoa'),\n",
       " ('Pectobacterium phage My1', 'Pectobacterium carotovorum subsp. carotovorum'),\n",
       " ('Strawberry necrotic shock virus', 'Fragaria'),\n",
       " ('Escherichia phage D108', 'Escherichia coli'),\n",
       " ('Lactococcus virus P2', 'Lactococcus lactis subsp. cremoris MG1363'),\n",
       " ('Yam mosaic virus', 'Nicotiana benthamiana'),\n",
       " ('Seal anellovirus 6', 'Phoca vitulina'),\n",
       " ('Escherichia phage UFV-AREG1', 'Escherichia coli O157'),\n",
       " ('Sweet potato leaf curl virus', 'Ipomoea alba'),\n",
       " ('Honeysuckle yellow vein virus', 'Lonicera japonica'),\n",
       " ('Saint Louis encephalitis virus', 'Dromaius novaehollandiae'),\n",
       " ('Sewage-associated circular DNA virus-22', 'root'),\n",
       " ('Wenzhou picorna-like virus 4', 'root'),\n",
       " ('Cymbidium ringspot virus satellite RNA',),\n",
       " ('Humulus japonicus latent virus', 'Chenopodium quinoa'),\n",
       " ('Gemycircularvirus HV-GcV1', 'Homo sapiens'),\n",
       " ('Leptospira phage vB_LalZ_80412-LE1',\n",
       "  'Leptospira alstonii serovar Pingchang str. 80-412'),\n",
       " ('Penaeid shrimp infectious myonecrosis virus', 'Penaeus vannamei'),\n",
       " ('Influenza C virus (C/Ann Arbor/1/50)', 'Homo sapiens'),\n",
       " ('Japanese encephalitis virus', 'Ardeidae'),\n",
       " ('Tomato chlorotic spot virus', 'Scirtothrips dorsalis'),\n",
       " ('Rhodococcus phage ReqiPoco6', 'Rhodococcus hoagii'),\n",
       " ('Hepatitis GB virus B', 'Callithrix jacchus'),\n",
       " ('Po-Circo-like virus 51', 'Sus scrofa'),\n",
       " ('Wenzhou qinvirus-like virus 2', 'root'),\n",
       " ('Malacosoma neustria nucleopolyhedrovirus', 'Lepidoptera'),\n",
       " ('Louping ill virus', 'Ovis aries'),\n",
       " ('Human rhinovirus B3', 'Homo sapiens'),\n",
       " ('Norovirus Hu/GII-4/Miyagi4/2006/JP', 'Homo sapiens'),\n",
       " ('White clover cryptic virus 2', 'Trifolium repens'),\n",
       " ('Khujand lyssavirus', 'Myotis mystacinus'),\n",
       " ('Feline astrovirus D1', 'Felis catus'),\n",
       " ('Sus scrofa papillomavirus 2', 'Sus scrofa'),\n",
       " ('Cymbidium mosaic virus', 'Cattleya'),\n",
       " ('Pseudomonas virus PaP3', 'Pseudomonas'),\n",
       " ('Changjiang picorna-like virus 6', 'root'),\n",
       " ('Streptomyces phage DrGrey', 'Streptomyces griseus'),\n",
       " ('Simian adenovirus B', 'Papio hamadryas'),\n",
       " ('Mycobacterium phage HufflyPuff', 'Mycolicibacterium smegmatis'),\n",
       " ('Brochothrix phage NF5', 'Brochothrix thermosphacta'),\n",
       " ('Giardia lamblia virus', 'Giardia intestinalis'),\n",
       " ('Squirrel adenovirus 1', 'Sciurus vulgaris'),\n",
       " ('Tobacco rattle virus', 'Hyacinthus'),\n",
       " ('Vicia cryptic virus', 'Vicia faba'),\n",
       " ('Pseudomonas phage Pf3', 'Pseudomonas aeruginosa'),\n",
       " ('Wuhan centipede virus', 'Otostigmus scaber'),\n",
       " ('Spring beauty latent virus', 'Gomphrena globosa'),\n",
       " ('Diatom colony associated ssRNA virus 2', 'Bacillariophyta'),\n",
       " ('Wenzhou shrimp virus 9', 'root'),\n",
       " ('Madariaga virus', 'Sapajus apella'),\n",
       " ('Camberwell virus', 'Homo sapiens'),\n",
       " ('Bacillus phage Pavlov', 'Bacillus megaterium'),\n",
       " ('Warrego virus',),\n",
       " ('Influenza A virus (A/swine/Pinar del Rio/3/2010(H1N1))',),\n",
       " ('Sodalis phage phiSG1', \"Sodalis glossinidius str. 'morsitans'\"),\n",
       " ('Spodoptera exigua iflavirus 2', 'Spodoptera exigua'),\n",
       " ('Mocis latipes granulovirus', 'Mocis'),\n",
       " ('Chimpanzee adenovirus Y25', 'Pan troglodytes'),\n",
       " ('Goat torovirus', 'Capra hircus'),\n",
       " ('Pseudomonas phage PaMx42', 'Pseudomonas aeruginosa'),\n",
       " ('Wenling zhaovirus-like virus 1', 'root'),\n",
       " ('Shahe picorna-like virus 2', 'root'),\n",
       " ('Rat cytomegalovirus ALL-03', 'Rattus rattus'),\n",
       " ('Beet mosaic virus', 'Amaranthus retroflexus'),\n",
       " ('Lactococcus virus ASCC281', 'Lactococcus lactis'),\n",
       " ('Gemycircularvirus sp.', 'Grus japonensis'),\n",
       " ('Ageratum enation virus associated DNA 1', 'Zinnia'),\n",
       " ('Lake Sarah-associated circular molecule 6', 'root'),\n",
       " ('Saimiri sciureus papillomavirus 3', 'Saimiri'),\n",
       " ('Lindernia anagallis yellow vein virus', 'Orobanchaceae'),\n",
       " ('Zucchini green mottle mosaic virus', 'Cucurbita pepo'),\n",
       " ('Pseudomonas phage DL68', 'Pseudomonas aeruginosa PAO1'),\n",
       " ('Potato mop-top virus', 'Solanum nigrum'),\n",
       " ('Bovine respiratory syncytial virus ATCC51908', 'Bos taurus'),\n",
       " ('Enterococcus phage vB_EfaS_IME198', 'Enterococcus faecalis'),\n",
       " ('Mungbean yellow mosaic India virus-[Mungbean]', 'Vigna radiata'),\n",
       " ('Murine norovirus GV/WU21/2005/USA', 'Mus'),\n",
       " ('Mycobacterium phage Lolly9', 'Mycolicibacterium smegmatis'),\n",
       " ('Wenzhou toti-like virus 1', 'root'),\n",
       " ('Maize streak virus - Reunion [N2AR2]', 'Zea mays subsp. mays'),\n",
       " ('Cotton leaf curl virus', 'Gossypium hirsutum'),\n",
       " ('Mycobacterium phage Damien', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Enterobacteria phage NC51', 'Enterobacterales'),\n",
       " ('Human rhinovirus B86', 'Homo sapiens'),\n",
       " ('Staphylococcus phage StauST398-1', 'Staphylococcus aureus'),\n",
       " ('Ranid herpesvirus 3', 'Rana temporaria'),\n",
       " ('Vernonia crinkle betasatellite', 'Gymnanthemum amygdalinum'),\n",
       " ('Ranid herpesvirus 1', 'Rana pipiens'),\n",
       " ('West African Asystasia virus 2', 'Asystasia gangetica'),\n",
       " ('Zalophus californianus papillomavirus 1', 'Zalophus californianus'),\n",
       " ('Erwinia phage vB_EamP-S6', 'Erwinia amylovora'),\n",
       " ('Avon-Heathcote Estuary associated circular virus 7',\n",
       "  'Austrovenus stutchburyi'),\n",
       " ('Capim virus', 'Caluromys'),\n",
       " ('Naranjal virus', 'Mesocricetus auratus'),\n",
       " ('Hubei lepidoptera virus 3', 'root'),\n",
       " ('Escherichia virus CBA120', 'Escherichia coli'),\n",
       " ('Halovirus HHTV-2', 'Haloarcula hispanica'),\n",
       " ('Acidianus rod-shaped virus 2', 'Sulfolobales'),\n",
       " ('Venezuelan equine encephalitis virus', 'Equus asinus'),\n",
       " ('Vibrio phage ICP1_2004_A', 'Vibrio cholerae'),\n",
       " ('Ecklonia radiata-associated virus 5', 'Ecklonia radiata'),\n",
       " ('Siegesbeckia yellow vein virus-[GD27]', 'Sigesbeckia glabrescens'),\n",
       " (\"Adoxophyes honmai entomopoxvirus 'L'\", 'Adoxophyes honmai'),\n",
       " ('Coleus blumei viroid 6', 'Plectranthus scutellarioides'),\n",
       " ('Norovirus Hu/GII-4/Saga1/2006/JP', 'Homo sapiens'),\n",
       " ('Ctenophore-associated circular virus 4', 'Mnemiopsis leidyi'),\n",
       " ('Hubei virga-like virus 15', 'root'),\n",
       " ('Barmah Forest virus', 'Culex annulirostris'),\n",
       " ('Norovirus Hu/Norwalk/20135/2009/VNM', 'Homo sapiens'),\n",
       " ('Lupine bocavirus', 'Canis lupus'),\n",
       " ('Wisteria vein mosaic virus', 'Wisteria'),\n",
       " ('Hubei picorna-like virus 29', 'root'),\n",
       " ('Corynebacterium phage Darwin', 'Corynebacterium vitaeruminis'),\n",
       " ('Mycobacterium virus Acadian', 'Mycolicibacterium smegmatis'),\n",
       " ('Coxsackievirus B3', 'Homo sapiens'),\n",
       " ('Mink coronavirus strain WD1127', 'Neovison vison'),\n",
       " ('Parvo-like hybrid virus UC11', 'Homo sapiens'),\n",
       " ('Xenotropic MuLV-related virus VP35', 'Homo sapiens'),\n",
       " ('Hubei leech virus 1', 'root'),\n",
       " ('Erectites yellow mosaic virus satellite DNA beta',\n",
       "  'Erechtites valerianifolius'),\n",
       " ('Mumps virus genotype G', 'Homo sapiens'),\n",
       " ('Acinetobacter phage vB_AbaM_ME3', 'Acinetobacter baumannii'),\n",
       " ('Human rhinovirus B5', 'Homo sapiens'),\n",
       " ('Mycobacterium phage Phlyer', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Shigella phage SHFML-26', 'Shigella sonnei'),\n",
       " ('Streptococcus phage M102AD', 'Streptococcus mutans'),\n",
       " ('Tobacco rattle virus', 'Cynara cardunculus var. scolymus'),\n",
       " ('Cattle blood-associated gemycircularvirus', 'Bos taurus'),\n",
       " ('Human adenovirus 46', 'Homo sapiens'),\n",
       " ('Squash mosaic virus', 'Cucurbita'),\n",
       " ('Hepatitis C virus subtype 2a', 'Homo sapiens'),\n",
       " ('Citrobacter phage CR44b', 'Citrobacter rodentium'),\n",
       " ('Human papillomavirus type 70', 'Homo sapiens'),\n",
       " ('Mirabilis mosaic virus', 'Mirabilis'),\n",
       " ('Xingshan nematode virus 6', 'root'),\n",
       " ('Staphylococcus phage StB20', 'Staphylococcus capitis'),\n",
       " ('Hepatitis C virus genotype 5', 'Homo sapiens'),\n",
       " ('Vibrio phage H2 SGB-2014', 'Vibrio cholerae'),\n",
       " ('Mycobacterium phage Gumball', 'Mycolicibacterium smegmatis'),\n",
       " ('Mycobacterium virus Bxb1', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Tomato yellow leaf curl Sardinia virus', 'Malva parviflora'),\n",
       " ('Pseudomonas phage PaoP5', 'Pseudomonas aeruginosa PAO1'),\n",
       " ('Bidens mottle virus', 'Helianthus annuus'),\n",
       " ('Senna leaf curl virus', 'Senna occidentalis'),\n",
       " ('Plantago lanceolata latent virus', 'Plantago lanceolata'),\n",
       " ('Achromobacter phage phiAxp-3', 'Achromobacter xylosoxidans'),\n",
       " ('Karang Sari virus', 'Culex vishnui'),\n",
       " ('Influenza A virus (A/swine/Holguin/121/2010(H1N1))', 'Gallus gallus'),\n",
       " ('Enterobacteria phage C-1 INW-2012', 'Enterobacteriaceae'),\n",
       " ('Maize streak virus - E[Pat]', 'Oryza sativa'),\n",
       " ('Sida yellow vein virus', 'Sida'),\n",
       " ('Bat adeno-associated virus YNM', 'Myotis ricketti'),\n",
       " ('Human rhinovirus B27', 'Homo sapiens'),\n",
       " ('Mink circovirus', 'Mustela lutreola'),\n",
       " ('Human papillomavirus type 13', 'Homo sapiens'),\n",
       " ('Staphylococcus phage StB27', 'Staphylococcus hominis'),\n",
       " ('Tobacco virus 1', 'Nicotiana tabacum'),\n",
       " ('Triumfetta yellow mosaic virus', 'Triumfetta'),\n",
       " ('Wenling narna-like virus 3', 'root'),\n",
       " ('Bat coronavirus HKU4-2', 'Chiroptera'),\n",
       " ('Pea enation mosaic virus 1', 'Cicer arietinum'),\n",
       " ('Changjiang narna-like virus 4', 'root'),\n",
       " ('Mirabilis jalapa mottle virus', 'Mirabilis jalapa'),\n",
       " ('Alternaria alternata chrysovirus 1', 'Alternaria alternata'),\n",
       " ('Mycobacterium phage Apizium', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Mycobacterium phage Turj99', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Maize streak virus - E[Pat]', 'Avena sativa'),\n",
       " ('Simian adenovirus 29', 'Homo sapiens'),\n",
       " ('Xenotropic MuLV-related virus 22Rv1/CWR-R1', 'Homo sapiens'),\n",
       " ('Bacillus virus 1', 'Bacillus sp. 6k512'),\n",
       " ('Magpie-robin coronavirus HKU18', 'Corvidae'),\n",
       " ('Strawberry mottle virus', 'Fragaria vesca'),\n",
       " ('Yellowstone lake phycodnavirus 2', 'root'),\n",
       " ('Acidianus filamentous virus 1', 'Acidianus infernus'),\n",
       " ('Tomato yellow leaf curl virus-[Jiroft:Iran]', 'Solanum lycopersicum'),\n",
       " ('Operophtera brumata nucleopolyhedrovirus', 'Operophtera brumata'),\n",
       " ('Trichoplusia ni ascovirus 2c', 'Noctuidae'),\n",
       " ('Pea enation mosaic virus 2', 'Pisum sativum'),\n",
       " ('Okra mottle virus', 'Abelmoschus esculentus'),\n",
       " ('Chickpea chlorosis Australia virus', 'Cicer arietinum'),\n",
       " ('Bokeloh bat lyssavirus', 'Myotis nattereri'),\n",
       " ('Bovine ephemeral fever virus', 'Bubalus bubalis'),\n",
       " ('Otarine picobirnavirus', 'Zalophus californianus'),\n",
       " ('Grapevine leafroll-associated virus 13', 'Vitis vinifera'),\n",
       " ('Alstroemeria virus X', 'Amaranthaceae'),\n",
       " ('Hubei sobemo-like virus 22', 'root'),\n",
       " ('Kedougou virus', 'Culicidae'),\n",
       " ('Yellow fever virus 17D/Tiantan', 'Homo sapiens'),\n",
       " ('Chenuda virus', 'Sus scrofa'),\n",
       " ('Gordonia phage Orchid', 'Gordonia terrae'),\n",
       " ('Parechovirus-like virus PLV-CHN', 'Homo sapiens'),\n",
       " ('Pseudomonas phage Pa2', 'Pseudomonas aeruginosa PAO1'),\n",
       " ('Lake Victoria marburgvirus - Angola2005', 'Chlorocebus aethiops'),\n",
       " ('Vibrio phage Vc1', 'Vibrio sp.'),\n",
       " ('Bovine viral diarrhea virus 3 Th/04_KhonKaen', 'Bos taurus'),\n",
       " ('Hop trefoil cryptic virus 2', 'Medicago lupulina'),\n",
       " ('Hubei tombus-like virus 26', 'root'),\n",
       " ('Human rhinovirus QCE', 'Homo sapiens'),\n",
       " ('Argentinian mammarenavirus', 'Akodon azarae'),\n",
       " ('Oat mosaic virus', 'Avena sativa'),\n",
       " ('Wuhan nido-like virus 1', 'root'),\n",
       " ('Biomphalaria virus 2', 'Biomphalaria'),\n",
       " ('Tonate virus',),\n",
       " ('Snake adenovirus 1', 'Pantherophis guttatus'),\n",
       " ('Okahandja mammarenavirus', 'Micaelamys namaquensis'),\n",
       " ('Wenzhou shrimp virus 3', 'root'),\n",
       " ('Human papillomavirus type 43', 'Homo sapiens'),\n",
       " ('Mycobacterium phage PBI1', 'Mycobacterium'),\n",
       " ('Norovirus Hu/GII-4/CGMH12/2007/TW', 'Homo sapiens'),\n",
       " ('West Nile virus SA93/01', 'Homo sapiens'),\n",
       " ('Gordonia phage GTE2', 'Nocardia brasiliensis'),\n",
       " ('Phytophthora infestans RNA virus 4', 'Phytophthora infestans'),\n",
       " ('Maize yellow dwarf virus-RMV2', 'Zea mays subsp. mays'),\n",
       " ('Velvet tobacco mottle virus Satellite RNA', 'Nicotiana clevelandii'),\n",
       " ('Escherichia phage ECML-4', 'Escherichia coli O157:H7'),\n",
       " ('Staphylococcus phage GRCS', 'Staphylococcus aureus'),\n",
       " ('Bean golden yellow mosaic virus-[Puerto Rico-Japan]', 'Phaseolus vulgaris'),\n",
       " ('Habenaria mosaic virus', 'Chenopodium quinoa'),\n",
       " ('Grapevine fanleaf virus satellite RNA', 'Chenopodium quinoa'),\n",
       " ('Alfalfa mosaic virus', 'Lupinus'),\n",
       " ('Choristoneura biennis entomopoxvirus', 'Choristoneura fumiferana'),\n",
       " ('Muromegalovirus G4', 'Mus musculus domesticus'),\n",
       " ('Camel associated drosmacovirus 2', 'Camelus dromedarius'),\n",
       " ('Torque teno mini virus 11', 'Homo sapiens'),\n",
       " ('Human astrovirus 8', 'Homo sapiens'),\n",
       " ('Bean dwarf mosaic virus', 'Glycine max'),\n",
       " ('Sewage-associated circular DNA virus-6', 'root'),\n",
       " ('Columbid circovirus', 'Columba livia'),\n",
       " ('Wenzhou picorna-like virus 24', 'root'),\n",
       " ('Diatom colony associated dsRNA virus 2', 'Bacillariophyta'),\n",
       " ('American plum line pattern virus', 'Prunus salicina'),\n",
       " ('Sanxia water strider virus 20', 'root'),\n",
       " ('Human mastadenovirus F', 'Homo sapiens'),\n",
       " ('Nylanderia fulva virus 1', 'Nylanderia fulva'),\n",
       " ('Lagos bat lyssavirus', 'Eidolon helvum'),\n",
       " ('Goose picornavirus 1', 'Anser albifrons'),\n",
       " ('Hollyhock leaf crumple virus-[Cairo]', 'Alcea rosea'),\n",
       " ('Melanoplus sanguinipes entomopoxvirus', 'Locusta migratoria'),\n",
       " ('Bacillus phage Grass', 'Bacillus subtilis'),\n",
       " ('Enterobacteria phage PR5', 'Enterobacterales'),\n",
       " ('Escherichia phage AR1', 'Escherichia coli O157:H7'),\n",
       " ('Lake Sarah-associated circular virus-2', 'root'),\n",
       " ('Fly associated circular virus 6', 'Diptera'),\n",
       " ('Human coxsackievirus A21 Coe', 'Homo sapiens'),\n",
       " ('Mycobacterium phage Numberten', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Human papillomavirus 4', 'Homo sapiens'),\n",
       " ('Apple hammerhead viroid-like circular RNA', 'Malus domestica'),\n",
       " (\"DeBrazza's monkey arterivirus\", 'Cercopithecus neglectus'),\n",
       " ('Sagiyama virus', 'Equus caballus'),\n",
       " ('Mycobacterium phage Astraea', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Pteropus associated gemycircularvirus 7', 'Pteropus tonganus'),\n",
       " ('Porcine stool-associated circular virus 5', 'Sus scrofa'),\n",
       " ('Tomato leaf curl China virus - OX2', 'Oxalis corniculata'),\n",
       " ('Feline coronavirus UU10', 'Felis catus'),\n",
       " ('Aphid lethal paralysis virus', 'Aphis'),\n",
       " ('Hubei tombus-like virus 6', 'root'),\n",
       " ('Mycobacterium phage Alex', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Propionibacterium phage PA1-14', 'Propionibacterium'),\n",
       " ('Sweet potato leaf curl virus', 'Ipomoea lacunosa'),\n",
       " ('Sinorhizobium phage phiLM21', 'Sinorhizobium sp. LM21'),\n",
       " ('Nesidiocoris tenuis virus', 'Nesidiocoris tenuis'),\n",
       " ('Pseudomonas phage Bjorn', 'Pseudomonas sp.'),\n",
       " ('Bos taurus papillomavirus 19', 'Bos taurus'),\n",
       " ('Elm mottle virus', 'Chenopodium quinoa'),\n",
       " ('Propionibacterium phage PHL009M11', 'Cutibacterium acnes'),\n",
       " ('Sanxia water strider virus 17', 'root'),\n",
       " ('Listeria phage A006', 'Listeria monocytogenes'),\n",
       " ('Vibrio phage VFJ', 'Vibrio cholerae O139'),\n",
       " ('Wheat yellow dwarf virus-GPV', 'Triticum aestivum'),\n",
       " ('Sinorhizobium phage phiN3', 'Sinorhizobium meliloti'),\n",
       " ('Lactococcus phage ASCC337', 'Lactococcus lactis'),\n",
       " ('Barfin flounder nervous necrosis virus', 'Gadus macrocephalus'),\n",
       " ('Ross River virus', 'Culex annulirostris'),\n",
       " ('Mycobacterium phage Tasp14', 'Mycolicibacterium smegmatis'),\n",
       " ('Bacillus phage Slash', 'Bacillus megaterium'),\n",
       " ('Cowpea mosaic virus', 'Vigna unguiculata'),\n",
       " ('Parry Creek virus', 'Culex annulirostris'),\n",
       " ('Bacillus phage Eyuki', 'Bacillus thuringiensis'),\n",
       " ('Mosqueiro virus', 'Culex portesi'),\n",
       " ('Rat coronavirus Parker', 'Rattus'),\n",
       " ('Hubei tombus-like virus 21', 'root'),\n",
       " ('Acholeplasma virus L2', 'Acholeplasma laidlawii'),\n",
       " ('Tomato yellow ring virus', 'Solanum lycopersicum'),\n",
       " ('Beihai picorna-like virus 77', 'root'),\n",
       " ('Maraba virus', 'Mus musculus'),\n",
       " ('Tomato black ring virus', 'Beta vulgaris'),\n",
       " ('Rice stripe tenuivirus', 'Setaria italica'),\n",
       " ('Acinetobacter phage AB3', 'Acinetobacter baumannii'),\n",
       " ('Bean yellow mosaic virus', 'Trifolium incarnatum'),\n",
       " ('Elderberry aureusvirus 1', 'Sambucus nigra'),\n",
       " ('Gemycircularvirus SL2', 'root'),\n",
       " ('Cellulophaga phage phi38:2', 'Cellulophaga baltica NN016038'),\n",
       " ('Potato yellow dwarf nucleorhabdovirus', 'Nicotiana rustica'),\n",
       " ('Arhar cryptic virus-I', 'Cajanus cajan'),\n",
       " ('Salmonella phage FSL SP-076',\n",
       "  'Salmonella enterica subsp. enterica serovar Dublin'),\n",
       " ('Silurus glanis circovirus', 'Silurus glanis'),\n",
       " ('Tomato yellow ring virus', 'Nicotiana benthamiana'),\n",
       " ('Tamana bat virus', 'Pteronotus parnellii'),\n",
       " ('Ageratum leaf curl Cameroon alphasatellite', 'Ageratum'),\n",
       " ('Maize rough dwarf virus', 'Triticum aestivum'),\n",
       " ('Sulfolobus spindle-shaped virus 2', 'Sulfolobus'),\n",
       " ('Black raspberry necrosis virus', 'Rubus occidentalis'),\n",
       " ('Barur virus', 'Rattus rattus'),\n",
       " ('Bat badicivirus 2', 'Eidolon helvum'),\n",
       " ('Mycobacterium phage PhatBacter', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Human papillomavirus type 10', 'Homo sapiens'),\n",
       " ('Puniceispirillum phage HMO-2011',\n",
       "  'Candidatus Puniceispirillum marinum IMCC1322'),\n",
       " ('Pseudoplusia includens SNPV IE', 'Chrysodeixis includens'),\n",
       " ('Haemophilus virus HP1', 'Haemophilus influenzae'),\n",
       " ('Circovirus-like genome DCCV-10', 'root'),\n",
       " ('Salivirus FHB', 'Homo sapiens'),\n",
       " ('Human parechovirus 8', 'Homo sapiens'),\n",
       " ('Tobacco mosaic virus', 'Nicotiana tabacum'),\n",
       " ('Duck megrivirus', 'Anatidae'),\n",
       " ('Mycobacterium phage Vista', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Apple green crinkle associated virus', 'Malus domestica'),\n",
       " ('Porcine endogenous retrovirus A', 'Sus scrofa'),\n",
       " ('Borna disease virus 1', 'Ovis aries'),\n",
       " ('Thermus phage phiYS40', 'Thermus thermophilus HB8'),\n",
       " ('Keterrah virus', 'Carios pusillus'),\n",
       " ('Satellites of Trichomonas vaginalis T1 virus', 'Trichomonas vaginalis'),\n",
       " ('Enhydra lutris papillomavirus 1', 'Enhydra lutris nereis'),\n",
       " ('Human papillomavirus type 68a', 'Homo sapiens'),\n",
       " ('Mycobacterium phage Omnicron', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Avian avulavirus 12', 'Aves'),\n",
       " ('Ntaya virus', 'Mus musculus'),\n",
       " ('Pseudomonas phage vB_PaeS_PM105', 'Pseudomonas aeruginosa'),\n",
       " ('Enterobacter phage phiEap-1', 'Enterobacter'),\n",
       " ('Rhinolophus simulator polyomavirus 3', 'Rhinolophus simulator'),\n",
       " ('Dolphin morbillivirus', 'Platanistidae'),\n",
       " ('Hirame novirhabdovirus', 'Plecoglossus altivelis'),\n",
       " ('Hubei tombus-like virus 18', 'root'),\n",
       " ('Salmonella phage SJ46',\n",
       "  'Salmonella enterica subsp. enterica serovar Indiana'),\n",
       " ('Sanxia sobemo-like virus 2', 'root'),\n",
       " ('Clostridium phage phiCT19406A', 'Clostridium tetani'),\n",
       " ('Fowl aviadenovirus E', 'Gallus gallus'),\n",
       " ('Hubei macula-like virus 1', 'root'),\n",
       " ('Enterococcus phage IME-EFm5', 'Enterococcus faecium'),\n",
       " ('American hop latent virus', 'Humulus lupulus'),\n",
       " ('Cotton leaf curl Alabad virus', 'Gossypium hirsutum'),\n",
       " ('Liberibacter phage FP2', 'Candidatus Liberibacter asiaticus str. psy62'),\n",
       " ('Norovirus Hu/GII.4/C00006631/2011/UK', 'Homo sapiens'),\n",
       " ('Aeromonas phage phiAS7', 'Aeromonas salmonicida subsp. salmonicida'),\n",
       " ('Staphylococcus phage GH15', 'Staphylococcus aureus'),\n",
       " ('Garlic virus A', 'Allium sativum'),\n",
       " ('Human papillomavirus type 11', 'Homo sapiens'),\n",
       " ('Synechococcus phage S-CBS2', 'Synechococcus sp. CB0204'),\n",
       " ('Feline immunodeficiency virus', 'Felidae'),\n",
       " ('Acute bee paralysis virus', 'Apis mellifera'),\n",
       " ('Chilli leaf curl Ahmedabad virus-India [India/Ahmedabad/2014]', 'Capsicum'),\n",
       " ('Andean potato mottle virus', 'Solanum tuberosum'),\n",
       " ('Norovirus Hu/Norwalk/20123/2009/VNM', 'Homo sapiens'),\n",
       " ('Eidolon helvum papillomavirus 2', 'Eidolon helvum'),\n",
       " ('Paenibacillus phage Willow', 'Paenibacillus larvae'),\n",
       " ('Mycobacterium phage Keshu', 'Mycolicibacterium smegmatis'),\n",
       " ('Abelson murine leukemia virus', 'Mus musculus'),\n",
       " ('Potato virus Y strain O', 'Capsicum'),\n",
       " ('Prunus necrotic ringspot virus', 'Chenopodium quinoa'),\n",
       " ('Rhodobacter phage RcapMu', 'Rhodobacter'),\n",
       " ('Bat SARS CoV Rm1/2004', 'Rhinolophus macrotis'),\n",
       " ('Dioscorea bacilliform RT virus 2', 'Dioscorea rotundata'),\n",
       " ('Chloris striate mosaic virus', 'Ixophorus unisetus'),\n",
       " ('Escherichia phage P483', 'Escherichia coli'),\n",
       " ('Hubei sobemo-like virus 10', 'root'),\n",
       " ('Mycobacterium phage Catera', 'Mycobacterium'),\n",
       " ('Rhynchobatus djiddensis adomavirus 1', 'Rhynchobatus djiddensis'),\n",
       " ('Shigella phage pSs-1', 'Shigella sonnei'),\n",
       " ('Tobacco vein distorting virus', 'Nicotiana tabacum'),\n",
       " ('Propionibacterium phage PAD20', 'Cutibacterium acnes'),\n",
       " ('Western chimpanzee simian foamy virus', 'Pan troglodytes'),\n",
       " ('African oil palm ringspot virus', 'Elaeis guineensis'),\n",
       " ('Enterobacteria phage NC3', 'Enterobacterales'),\n",
       " ('Dragonfly larvae associated circular virus-3', 'Procordulia grayi'),\n",
       " ('Mycobacterium phage Breeniome', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Sparus aurata polyomavirus 1', 'Sparus aurata'),\n",
       " ('Human mastadenovirus B', 'Homo sapiens'),\n",
       " ('Brochothrix phage A9', 'Brochothrix thermosphacta'),\n",
       " ('Wuhan spider virus 6', 'root'),\n",
       " ('Tomato pseudo-curly top virus', 'Solanum nigrum'),\n",
       " ('Blackcurrant reversion virus satellite RNA',),\n",
       " ('Dengue virus 2', 'Aedes aegypti'),\n",
       " ('Pennisetum mosaic virus', 'Setaria'),\n",
       " ('Escherichia phage mEpX2', 'Escherichia coli'),\n",
       " ('Listonella phage phiHSIC', 'Vibrio'),\n",
       " ('San Miguel sea lion virus', 'Zalophus californianus'),\n",
       " ('Soil-borne wheat mosaic virus', 'Hordeum vulgare'),\n",
       " ('Wenzhou bivalvia virus 2', 'root'),\n",
       " ('Rice black-streaked dwarf virus 2', 'Oryza sativa'),\n",
       " ('Grapevine chrome mosaic virus', 'Apium graveolens'),\n",
       " ('Sewage-associated circular DNA virus-26', 'root'),\n",
       " ('Acanthamoeba polyphaga moumouvirus', 'Acanthamoeba polyphaga'),\n",
       " ('Escherichia coli O157 typing phage 12', 'Escherichia coli O157:H7'),\n",
       " ('Mycobacterium phage Fishburne', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Haloarcula hispanica pleomorphic virus 2',\n",
       "  'Haloarcula hispanica ATCC 33960'),\n",
       " ('Bovine gammaherpesvirus 4', 'Felis catus'),\n",
       " ('Potato virus Y strain O', 'Solanum tuberosum'),\n",
       " ('Maize streak virus - [Raw]', 'Brachiaria deflexa'),\n",
       " ('Lactococcus phage ASCC365', 'Lactococcus lactis'),\n",
       " ('Rehmannia mosaic virus', 'Rehmannia glutinosa'),\n",
       " ('Ilheus virus', 'Haemagogus'),\n",
       " ('Ageratum yellow vein betasatellite', 'Nicotiana benthamiana'),\n",
       " ('Sierra dome spider associated circular virus 2',),\n",
       " ('Clostridium phage phiCD111', 'Clostridium'),\n",
       " ('Goatpox virus Pellor', 'Capra hircus'),\n",
       " ('Common bottlenose dolphin gammaherpesvirus 1 strain Sarasota',\n",
       "  'Tursiops truncatus'),\n",
       " ('Norovirus Hu/GII.4/C00007927/2010/UK', 'Homo sapiens'),\n",
       " ('Colorado tick fever virus', 'Dermacentor andersoni'),\n",
       " ('Potato virus Y strain Wilga 156var', 'Solanum tuberosum'),\n",
       " ('Apple stem pitting virus', 'Pyrus communis'),\n",
       " ('Choristoneura murinana nucleopolyhedrovirus', 'Cydia pomonella'),\n",
       " ('African eggplant yellowing virus', 'Solanum aethiopicum'),\n",
       " ('Mimosa yellow leaf curl virus-associated DNA 1', 'Fabaceae'),\n",
       " ('Pepper mild mottle virus', 'Vicia faba'),\n",
       " ('Andes orthohantavirus', 'Oligoryzomys longicaudatus'),\n",
       " ('Baminivirus', 'root'),\n",
       " ('Mycobacterium phage Send513', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Mycobacterium phage Dhanush', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Mycobacterium virus Dorothy', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Staphylococcus phage SA97', 'Staphylococcus aureus subsp. aureus RN4220'),\n",
       " ('Beihai picorna-like virus 70', 'root'),\n",
       " ('Human adenovirus 28', 'Homo sapiens'),\n",
       " ('Lisianthus necrosis virus', 'Eustoma exaltatum subsp. russellianum'),\n",
       " ('Murine norovirus GV/WU24/2005/USA', 'Mus'),\n",
       " ('Blainvillea yellow spot virus', 'Blainvillea rhomboidea'),\n",
       " ('Mycobacterium phage Sbash', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Rice dwarf virus', 'Oryza sativa'),\n",
       " ('Staphylococcus virus 29', 'Staphylococcus aureus'),\n",
       " ('Human polyomavirus 7', 'Homo sapiens'),\n",
       " ('Equine pegivirus 1', 'Equus caballus'),\n",
       " ('Xincheng Mosquito Virus', 'Anopheles sinensis'),\n",
       " ('Rinderpest virus (strain Kabete O)', 'Capra hircus'),\n",
       " ('Lactobacillus phage iLp1308', 'Lactobacillus paracasei'),\n",
       " ('Lactococcus lactis phage 645', 'Lactococcus lactis'),\n",
       " ('Rice yellow stunt nucleorhabdovirus', 'Oryza sativa'),\n",
       " ('Melon yellow spot virus', 'Gomphrena globosa'),\n",
       " ('Mannheimia phage vB_MhS_1152AP2', 'Mannheimia haemolytica'),\n",
       " ('Mycobacterium phage Adler', 'Mycobacterium'),\n",
       " ('McMurdo Ice Shelf pond-associated circular DNA virus-6', 'root'),\n",
       " ('Molossus molossus circovirus 1', 'Molossus molossus'),\n",
       " ('Mycobacterium phage Dreamboat', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Buzura suppressaria nucleopolyhedrovirus', 'Lepidoptera'),\n",
       " ('Pyrobaculum spherical virus', 'Pyrobaculum sp. D11'),\n",
       " ('Bovine enterovirus type 2', 'Bos taurus'),\n",
       " ('Murine hepatitis virus strain A59', 'Mus'),\n",
       " ('Orangutan hepadnavirus', 'Pongo abelii'),\n",
       " ('Winged bean alphaendornavirus 1', 'Psophocarpus tetragonolobus'),\n",
       " ('Lactococcus phage bIL311', 'Lactococcus lactis subsp. lactis Il1403'),\n",
       " ('Norovirus Hu/GII-4/Kumamoto3/2006/JP', 'Homo sapiens'),\n",
       " ('Karshi virus', 'Chlorocebus aethiops'),\n",
       " ('Beet mosaic virus', 'Sonchus arvensis'),\n",
       " ('Lymphocystis disease virus Sa', 'Sparus aurata'),\n",
       " ('Israeli acute paralysis virus', 'Apis mellifera'),\n",
       " ('Choristoneura fumiferana granulovirus', 'Choristoneura fumiferana'),\n",
       " ('Bat rotavirus', 'Rousettus aegyptiacus'),\n",
       " ('Changjiang tombus-like virus 9', 'root'),\n",
       " ('Hubei picorna-like virus 8', 'root'),\n",
       " ('Bacillus phage Pookie', 'Bacillus megaterium'),\n",
       " ('Okra leaf curl alphasatellite', 'Okra yellow crinkle virus'),\n",
       " ('uncultured crAssphage', 'bacterium'),\n",
       " ('Tomato chlorosis virus', 'Solanum lycopersicum'),\n",
       " ('Dendrolimus punctatus densovirus', 'Dendrolimus punctatus'),\n",
       " ('Staphylococcus phage phiSA12', 'Staphylococcus aureus'),\n",
       " ('Marine RNA virus JP-B', 'root'),\n",
       " ('Thermus virus P23-77', 'Thermus thermophilus'),\n",
       " ('Astrovirus SG', 'Homo sapiens'),\n",
       " ('Mycobacterium phage Vortex', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Orchid fleck dichorhavirus', 'Phalaenopsis'),\n",
       " ('Papaya leaf crumple virus-Nirulas [India:New Delhi:Papaya:2007]',\n",
       "  'Carica papaya'),\n",
       " ('Sanxia tombus-like virus 1', 'root'),\n",
       " ('Soybean latent spherical virus', 'Glycine max'),\n",
       " ('Odonata-associated circular virus-3', 'Ischnura ramburii'),\n",
       " ('Bombyx mori densovirus 5', 'Bombyx mori'),\n",
       " ('Mycobacterium phage ArcherNM', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Bat SARS coronavirus HKU3-12', 'Chiroptera'),\n",
       " ('Beihai narna-like virus 11', 'root'),\n",
       " ('Dugbe orthonairovirus', 'Rhipicephalus'),\n",
       " ('Microbacterium phage Koji', 'Microbacterium foliorum'),\n",
       " ('Ovine lentivirus', 'Ovis aries'),\n",
       " ('Squash vein yellowing virus', 'Citrullus lanatus subsp. vulgaris'),\n",
       " ('Dioscorea bacilliform virus', 'Dioscorea sansibarensis'),\n",
       " ('Nienokoue virus', 'Culex'),\n",
       " ('Wheat dwarf India virus', 'Triticum aestivum'),\n",
       " ('Hendra henipavirus', 'Pteropus scapulatus'),\n",
       " ('Bat CoV 273/2005',),\n",
       " ('Norovirus Hu/GII.12/CGMH40/2010/TW', 'Homo sapiens'),\n",
       " ('Staphylococcus virus 13', 'Staphylococcus aureus subsp. aureus NCTC 8325'),\n",
       " ('Flock House virus', 'Galleria mellonella'),\n",
       " ('Nodamura virus', 'Culex tarsalis'),\n",
       " ('Cowpox virus', 'Rhombomys opimus'),\n",
       " ('Pepper enamovirus', 'Capsicum'),\n",
       " ('Echovirus E15', 'Homo sapiens'),\n",
       " ('Sporobolus striate mosaic virus 2', 'Sporobolus'),\n",
       " ('Bos grunniens papillomavirus type 1', 'Bos grunniens'),\n",
       " ('Arachis pintoi virus', 'Arachis pintoi'),\n",
       " ('Cowpea golden mosaic virus-[Nigeria]', 'Vigna unguiculata'),\n",
       " ('Mycobacterium virus Jasper', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Ustilaginoidea virens RNA virus 3', 'Ustilaginoidea virens'),\n",
       " ('Black grass cryptic virus 2', 'Alopecurus myosuroides'),\n",
       " ('Mycobacterium phage Brocalys', 'Mycolicibacterium smegmatis'),\n",
       " ('Human papillomavirus 140', 'Homo sapiens'),\n",
       " ('Simian torque teno virus 31', 'Chlorocebus aethiops'),\n",
       " ('HBV recombinant C/G', 'Homo sapiens'),\n",
       " ('Mumps virus strain Jeryl Lynn', 'Homo sapiens'),\n",
       " ('Circoviridae SFBeef', 'Bos taurus'),\n",
       " ('Beihai sobemo-like virus 2', 'root'),\n",
       " ('African swine fever virus Benin 97/1', 'Phacochoerus africanus'),\n",
       " ('Privet leaf blotch-associated virus', 'Ligustrum japonicum'),\n",
       " ('Macaca mulatta papillomavirus 3', 'Macaca mulatta'),\n",
       " ('Pseudomonas phage MPK6', 'Pseudomonas aeruginosa'),\n",
       " ('Mycobacterium phage PattyP', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Beihai sobemo-like virus 3', 'root'),\n",
       " ('Mouse parvovirus 1e', 'Mus musculus'),\n",
       " ('Culex flavivirus', 'Culex pipiens'),\n",
       " ('Odonata-associated circular virus-16', 'Libellula quadrimaculata'),\n",
       " ('Ralstonia virus RSA1', 'Ralstonia solanacearum'),\n",
       " ('Brucella phage F1', 'Brucella sp.'),\n",
       " ('Halorubrum pleomorphic virus 2', 'Halorubrum sp. SS5-4'),\n",
       " ('Xanthophyllomyces dendrorhous virus L1A', 'Xanthophyllomyces dendrorhous'),\n",
       " ('Escherichia phage vB_EcoP_PhAPEC7', 'Escherichia coli'),\n",
       " ('Sacbrood virus', 'Apis mellifera'),\n",
       " ('Adoxophyes orana nucleopolyhedrovirus', 'Adoxophyes orana'),\n",
       " ('Merremia mosaic Puerto Rico virus', 'Distimake quinquefolius'),\n",
       " ('Torque teno sus virus 1b', 'Sus scrofa'),\n",
       " ('Hubei odonate virus 11', 'root'),\n",
       " ('Spiroplasma phage 1-C74', 'Spiroplasma melliferum'),\n",
       " ('Tomato golden mosaic virus', 'Physalis'),\n",
       " ('Beihai sipunculid worm virus 4', 'root'),\n",
       " ('Porcine rubulavirus', 'Sus scrofa'),\n",
       " ('Coleus blumei viroid', 'Plectranthus scutellarioides'),\n",
       " ('Erethizon dorsatum papillomavirus 2', 'Erethizon dorsatum'),\n",
       " ('Bean golden yellow mosaic virus-[Puerto Rico-Japan]',\n",
       "  'Macroptilium lathyroides'),\n",
       " ('Mycobacterium phage Trike', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Wenzhou picorna-like virus 2', 'root'),\n",
       " ('Thin paspalum asymptomatic virus', 'Paspalum setaceum'),\n",
       " ('Human papillomavirus type 56', 'Homo sapiens'),\n",
       " ('Wuhan cricket virus', 'Conocephalus'),\n",
       " ('Cucurbit leaf crumple virus', 'Citrullus lanatus'),\n",
       " ('Melon yellow spot virus', 'Lagenaria siceraria'),\n",
       " ('Bemisia-associated genomovirus AdO', 'Bemisia tabaci'),\n",
       " ('Southwest carpet python virus', 'Morelia spilota'),\n",
       " ('Arctopus echinatus-associated virus', 'Arctopus echinatus'),\n",
       " ('Rhinolophus gammaherpesvirus 1', 'Rhinolophus ferrumequinum'),\n",
       " ('Epizootic hemorrhagic disease virus (serotype 5 / strain CSIRO 157)',\n",
       "  'Bos taurus'),\n",
       " ('Rhesus adenovirus 53', 'Macaca mulatta'),\n",
       " ('Salmonella phage Fels-1',\n",
       "  'Salmonella enterica subsp. enterica serovar Typhimurium str. LT2'),\n",
       " ('Pedilanthus leaf curl virus', 'Pedilanthus'),\n",
       " ('Erwinia phage vB_EamM_RAY', 'Erwinia amylovora'),\n",
       " ('Tomato yellow vein streak virus', 'Solanum lycopersicum'),\n",
       " ('Xanthomonas phage OP2', 'Xanthomonas'),\n",
       " ('Radish leaf curl virus', 'Raphanus sativus'),\n",
       " ('SARS coronavirus PUMC01', 'Mus musculus'),\n",
       " ('Porcine circovirus', 'Sus scrofa'),\n",
       " ('Wenling crustacean virus 2', 'root'),\n",
       " ('Torque teno virus 10', 'Homo sapiens'),\n",
       " ('Alfalfa dwarf virus', 'Medicago sativa'),\n",
       " ('Escherichia virus Jk06', 'Escherichia'),\n",
       " ('Propionibacterium phage PHL141N00', 'Cutibacterium acnes'),\n",
       " ('Decapod penstyldensovirus 1', 'Penaeidae'),\n",
       " ('Hubei unio douglasiae virus 3', 'root'),\n",
       " ('Pepper huasteco yellow vein virus', 'Capsicum'),\n",
       " ('Beet mild yellowing virus', 'Capsella bursa-pastoris'),\n",
       " ('Human papillomavirus type 87', 'Homo sapiens'),\n",
       " ('Human parainfluenza virus 4b', 'Homo sapiens'),\n",
       " ('Plantago mottle virus', 'Tetragonia tetragonoides'),\n",
       " ('Mycobacterium phage 20ES', 'Mycolicibacterium smegmatis'),\n",
       " ('Xenotropic MuLV-related virus RKO', 'Homo sapiens'),\n",
       " ('Pseudomonas phage vB_PaeM_C1-14_Ab28', 'Pseudomonas aeruginosa'),\n",
       " ('Human papillomavirus type 147', 'Homo sapiens'),\n",
       " ('Norovirus Hu/GII/BG1C0066/2011/BGD', 'Homo sapiens'),\n",
       " ('Pan troglodytes polyomavirus 3', 'Pan troglodytes verus'),\n",
       " ('Streptococcus phage K13', 'Streptococcus pneumoniae'),\n",
       " ('Porcine stool-associated circular virus/BEL/15V010', 'Sus scrofa'),\n",
       " ('Human papillomavirus type 103', 'Homo sapiens'),\n",
       " ('Wenzhou narna-like virus 8', 'root'),\n",
       " ('Agrotis segetum nucleopolyhedrovirus A', 'Lepidoptera'),\n",
       " ('Hubei picorna-like virus 48', 'root'),\n",
       " ('Clostridium phage phiMMP03', 'Clostridium'),\n",
       " ('Enterococcus phage SAP6', 'Enterococcus faecalis'),\n",
       " ('Picornavirales Bu-1', 'Sus scrofa domesticus'),\n",
       " ('Ilheus virus', 'Trichoprosopon'),\n",
       " ('Mungbean yellow mosaic India virus', 'Vigna mungo'),\n",
       " ('Rusa timorensis papillomavirus type 1', 'Rusa timorensis'),\n",
       " ('Cyanophage S-RIM50', 'Synechococcus sp.'),\n",
       " ('Beet western yellows ST9 associated virus', 'Capsella bursa-pastoris'),\n",
       " ('Changjiang sobemo-like virus 1', 'root'),\n",
       " ('Pseudomonas phage vB_PaeM_PAO1_Ab29', 'Pseudomonas aeruginosa'),\n",
       " ('Mycobacterium phage Emerson', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Kenaf leaf curl betasatellite', 'Alcea rosea'),\n",
       " ('Salmonella phage f18SE', 'Salmonella enterica'),\n",
       " ('Bovine faeces associated smacovirus 2', 'Bos taurus'),\n",
       " ('Propionibacterium phage PHL092M00', 'Cutibacterium acnes'),\n",
       " ('Mycobacterium phage Bipolar', 'Mycolicibacterium smegmatis'),\n",
       " ('Coxsackievirus A17', 'Homo sapiens'),\n",
       " ('Cotia virus', 'Chlorocebus aethiops'),\n",
       " ('Enterobacteria phage If1', 'Escherichia coli'),\n",
       " ('Eptesicus serotinus papillomavirus 1', 'Eptesicus serotinus'),\n",
       " ('Emilia yellow vein virus-associated DNA beta', 'Emilia sonchifolia'),\n",
       " ('Hubei picorna-like virus 31', 'root'),\n",
       " ('Linda virus', 'Sus scrofa'),\n",
       " ('Polytropic murine leukemia virus', 'Mus musculus'),\n",
       " ('Porcine enterovirus 9', 'Sus scrofa'),\n",
       " ('Mycobacterium phage Medusa', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Pine marten torque teno virus 1', 'Martes martes'),\n",
       " ('Honeysuckle yellow vein virus-[UK2]', 'Lonicera japonica'),\n",
       " ('Aedes camptorhynchus toti-like virus 1', 'Aedes'),\n",
       " ('Itacaiunas virus', 'Culicoides'),\n",
       " ('Sorghum mosaic virus', 'Saccharum hybrid cultivar'),\n",
       " ('Delftia phage IME-DE1', 'Delftia sp. 670'),\n",
       " ('Clover yellow vein virus', 'Lupinus luteus'),\n",
       " ('Urucuri virus', 'Proechimys guyannensis'),\n",
       " ('Isaria javanica chrysovirus 1', 'Cordyceps javanica'),\n",
       " ('Cebus albifrons polyomavirus 1', 'Cebus albifrons'),\n",
       " ('Beihai sobemo-like virus 4', 'root'),\n",
       " ('Porcine parvovirus', 'Sus scrofa'),\n",
       " ('Propionibacterium phage PHL113M01', 'Cutibacterium acnes'),\n",
       " ('Vibrio phage ICP1', 'Vibrio cholerae'),\n",
       " ('Sweet potato vein mosaic virus', 'Ipomoea batatas'),\n",
       " ('Carrot Ch virus 2', 'Daucus carota'),\n",
       " ('Wesselsbron virus', 'Homo sapiens'),\n",
       " ('SARS coronavirus WHU', 'Homo sapiens'),\n",
       " ('Tobacco yellow dwarf virus', 'Solanum lycopersicum'),\n",
       " ('Mobuck virus', 'Odocoileus virginianus'),\n",
       " ('Ferret coronavirus', 'Mustela putorius furo'),\n",
       " ('Parsnip yellow fleck virus', 'Anethum graveolens'),\n",
       " ('Mycobacterium virus Liefie', 'Mycolicibacterium smegmatis'),\n",
       " ('Leuconostoc phage Ln-9', 'Leuconostoc mesenteroides subsp. mesenteroides'),\n",
       " ('Salmonella phage SSE121', 'Salmonella'),\n",
       " ('Sodalis phage SO1', 'Sodalis glossinidius'),\n",
       " ('Synedrella leaf curl alphasatellite', 'Synedrella nodiflora'),\n",
       " ('Uukuniemi virus', 'Ixodes ricinus'),\n",
       " ('Lettuce yellow mottle virus', 'Datura stramonium'),\n",
       " ('Zucchini green mottle mosaic virus', 'Lagenaria siceraria'),\n",
       " ('Bean calico mosaic virus', 'Phaseolus acutifolius'),\n",
       " ('Jingmen tick virus', 'Rhipicephalus microplus'),\n",
       " ('Papaya meleira virus', 'Carica papaya'),\n",
       " ('Cyclovirus NGchicken15/NGA/2009', 'Gallus gallus'),\n",
       " ('Bean leafroll virus', 'Vicia faba'),\n",
       " ('Duck faeces associated circular DNA virus 1', 'Anas platyrhynchos'),\n",
       " ('Parsnip yellow fleck virus', 'Chaerophyllum temulum'),\n",
       " ('Sinorhizobium phage phiM9', 'Sinorhizobium meliloti'),\n",
       " ('Mycobacterium phage NelitzaMV', 'Mycolicibacterium smegmatis MC2 155'),\n",
       " ('Sphaeropsis sapinea RNA virus 2', 'Diplodia sapinea'),\n",
       " ('Human papillomavirus 159', 'Homo sapiens'),\n",
       " ('Minute virus of mice', 'Mus musculus'),\n",
       " ('Norovirus GII/Hu/JP/2012/GII.Pe_GII.4_Sydney2012/Fukuyama/1',\n",
       "  'Homo sapiens'),\n",
       " ('Mycobacterium phage Tiffany', 'Mycobacterium'),\n",
       " ('Pseudomonas phage F8', 'Pseudomonas aeruginosa'),\n",
       " ('Wenling tombus-like virus 5', 'root'),\n",
       " ('Chiltepin yellow mosaic virus', 'Capsicum annuum var. glabriusculum'),\n",
       " ('Paspalum striate mosaic virus', 'Paspalum dilatatum'),\n",
       " ('Porcine picobirnavirus', 'Sus scrofa'),\n",
       " ('Pegivirus F', 'Carollia perspicillata'),\n",
       " ('Pseudomonas virus MP22', 'Pseudomonas aeruginosa'),\n",
       " ('Cestrum yellow leaf curling virus', 'Cestrum parqui'),\n",
       " ('Pea enation mosaic virus 1', 'Trifolium incarnatum'),\n",
       " ('Dickeya phage RC-2014', 'Dickeya sp.'),\n",
       " ('Simian torque teno virus 34', 'Chlorocebus aethiops'),\n",
       " ('Bean yellow mosaic virus', 'Lupinus luteus'),\n",
       " ('Feline coronavirus UU8', 'Felis catus'),\n",
       " ('Staphylococcus phage SPbeta-like', 'Staphylococcus epidermidis 36-1'),\n",
       " ('Barley yellow dwarf virus PAV', 'Hordeum vulgare'),\n",
       " ('Malvastrum yellow vein virus-[Y47]', 'Malvastrum'),\n",
       " ('Budgerigar fledgling disease virus - 1', 'Melopsittacus undulatus'),\n",
       " ('Salmonid alphavirus subtype 3', 'Salmo salar'),\n",
       " ('Hubei qinvirus-like virus 1', 'root'),\n",
       " ('Maize streak virus - [Raw]', 'Oryza glaberrima'),\n",
       " ('Beihai picorna-like virus 6', 'root'),\n",
       " ('Tomato leaf curl Sudan virus', 'Phaseolus vulgaris'),\n",
       " ('Botrytis cinerea fusarivirus 1-S1', 'Botrytis cinerea'),\n",
       " ('Xenotropic murine leukemia virus', 'Homo sapiens'),\n",
       " ('Norovirus GII/Hu/JP/2007/GII.P7_GII.14/Sendai/YG99', 'Homo sapiens'),\n",
       " ('Passerivirus A1', 'Turdus pallidus'),\n",
       " ('Wenzhou picorna-like virus 17', 'root'),\n",
       " ('Circovirus-like genome DCCV-3', 'root'),\n",
       " ('Corchorus yellow vein mosaic betasatellite', 'Corchorus olitorius'),\n",
       " ('Lactobacillus phage LfeSau', 'Lactobacillus fermentum'),\n",
       " ('Beihai tombus-like virus 1', 'root'),\n",
       " ('Mycobacterium phage Orion', 'Mycolicibacterium smegmatis'),\n",
       " ('Pan troglodytes verus polyomavirus 2c', 'Pan troglodytes verus'),\n",
       " ('Norovirus Hu/GII.3/CBNU1/2006/KOR', 'Homo sapiens'),\n",
       " ('Propionibacterium phage PHL163M00', 'Cutibacterium acnes'),\n",
       " ('Rockport virus', 'Scalopus aquaticus'),\n",
       " ('Acinetobacter phage Acj61', 'Acinetobacter johnsonii'),\n",
       " ('Usutu virus', 'Culex perexiguus'),\n",
       " ('Euphorbia yellow mosaic virus', 'Euphorbia heterophylla'),\n",
       " ('Shahe arthropod virus 1', 'root'),\n",
       " ('Sulfolobus polyhedral virus 1', 'Sulfolobus sp.'),\n",
       " ('Transmissible gastroenteritis virus', 'Sus scrofa'),\n",
       " ('Canis familiaris papillomavirus 15', 'Canis lupus familiaris'),\n",
       " ('Ceratobasidium endornavirus A', 'Ceratobasidium sp.'),\n",
       " ('Human polyomavirus 1', 'Homo sapiens'),\n",
       " ('Odonata-associated circular virus-20', 'Erythemis simplicicollis'),\n",
       " ('Influenza A virus (A/Puerto Rico/8/34/Mount Sinai(H1N1))', 'Homo sapiens'),\n",
       " ('Sea otter poxvirus', 'Enhydra lutris kenyoni'),\n",
       " ('Synechococcus phage S-MbCM100', 'Synechococcus sp. WH 7803'),\n",
       " ('Tomato leaf curl Sulawesi virus', 'Solanum lycopersicum'),\n",
       " ('Feline coronavirus UU23', 'Felis'),\n",
       " ('Ockelbo virus', 'Acrocephalus scirpaceus'),\n",
       " ('Avian infectious bronchitis virus partridge/GD/S14/2003', 'Aves'),\n",
       " ('Staphylococcus phage S25-4', 'Staphylococcus aureus'),\n",
       " ('Human astrovirus 5', 'Homo sapiens'),\n",
       " ('Norovirus GII/Hu/JP/2001/GII.P22_GII.5/Saitama/T49', 'Homo sapiens'),\n",
       " ('Propionibacterium phage PHL117M01', 'Cutibacterium acnes'),\n",
       " ('Hubei picorna-like virus 44', 'root'),\n",
       " ('Molossus molossus circovirus 2', 'Molossus molossus'),\n",
       " ('Klebsiella phage K64-1', 'Klebsiella pneumoniae'),\n",
       " ('Mycobacterium virus CJW1', 'Mycobacterium'),\n",
       " ('Pepper yellow mosaic virus', 'Capsicum annuum'),\n",
       " ('Gull circovirus', 'Ichthyaetus leucophthalmus'),\n",
       " ('Norovirus Hu/GIV.1/LakeMacquarie/NSW268O/2010/AU', 'Homo sapiens'),\n",
       " ('Beet necrotic yellow vein virus', 'Beta vulgaris'),\n",
       " ('Corvus monedula polyomavirus 1', 'Corvus monedula'),\n",
       " ('Tobacco leaf curl PUSA alphasatellite', 'Nicotiana tabacum'),\n",
       " ('European bat 2 lyssavirus', 'Homo sapiens'),\n",
       " ('Bat coronavirus HKU4-4', 'Chiroptera'),\n",
       " ('Cotton chlorotic spot virus', 'Gossypium hirsutum'),\n",
       " ('Pseudomonas virus LUZ24', 'Pseudomonas aeruginosa PAO1'),\n",
       " ('Cauliflower mosaic virus', 'Raphanus'),\n",
       " ('Thermoproteus tenax virus 1', 'Thermoproteus tenax'),\n",
       " ('Gallivirus Pf-CHK1/GV', 'Gallus gallus'),\n",
       " ('Simian T-cell lymphotropic virus 4', 'Gorilla gorilla'),\n",
       " ('Enterobacteria phage DT571/2', 'Escherichia coli'),\n",
       " ('Ludwigia yellow vein virus', 'Ludwigia hyssopifolia'),\n",
       " ('Avian adeno-associated virus', 'Gallus gallus'),\n",
       " ('Hubei picorna-like virus 75', 'root'),\n",
       " ('Mycobacterium phage Lilac', 'Mycolicibacterium smegmatis'),\n",
       " ('Tete orthobunyavirus', 'Ploceus cucullatus'),\n",
       " ('Adelie penguin polyomavirus', 'Pygoscelis adeliae'),\n",
       " ('Bovine papillomavirus', 'Bos taurus'),\n",
       " ('Maize rough dwarf virus', 'Zea mays'),\n",
       " ('Macaca nemestrina herpesvirus 7', 'Macaca nemestrina'),\n",
       " ('Cotton leaf curl Gezira virus', 'Abelmoschus esculentus'),\n",
       " ('Macaca fascicularis papillomavirus 10', 'Macaca fascicularis'),\n",
       " ...}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "known_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of all LFs\n",
    "LFs = [\n",
    "     LF_detect, LF_detect2, LF_infect, LF_isolate, LF_positive, LF_positive2, LF_misc, LF_v_cause_h, LF_v_h, LF_h_v, LF_other_verbs, LF_uncertain, LF_far_v_h, LF_far_h_v, LF_neg_h, LF_distant_supervision, LF_neg_assertions\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number labeled: 3397\n"
     ]
    }
   ],
   "source": [
    "# To label and view LFs for testing\n",
    "labeled = []\n",
    "for c in session.query(VirusHost).filter(VirusHost.split == 0).all():\n",
    "    for function in LFs:\n",
    "        if function(c) != 0:\n",
    "            if c not in labeled:\n",
    "                labeled.append(c)\n",
    "print(\"Number labeled:\", len(labeled))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "##### To view only a specific LF\n",
    "\n",
    "labeled2 = []\n",
    "for c in session.query(VirusHost).filter(VirusHost.split == 0).all():\n",
    "    if LF_far_v_h(c) != 0: # change this line to test a specified LF\n",
    "        if c not in labeled2:\n",
    "            labeled2.append(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "require.undef('viewer');\n",
       "\n",
       "// NOTE: all elements should be selected using this.$el.find to avoid collisions with other Viewers\n",
       "\n",
       "define('viewer', [\"@jupyter-widgets/base\"], function(widgets) {\n",
       "    var ViewerView = widgets.DOMWidgetView.extend({\n",
       "        render: function() {\n",
       "            this.cids   = this.model.get('cids');\n",
       "            this.nPages = this.cids.length;\n",
       "            this.pid  = 0;\n",
       "            this.cxid = 0;\n",
       "            this.cid  = 0;\n",
       "\n",
       "            // Insert the html payload\n",
       "            this.$el.append(this.model.get('html'));\n",
       "\n",
       "            // Initialize all labels from previous sessions\n",
       "            this.labels = this.deserializeDict(this.model.get('_labels_serialized'));\n",
       "            for (var i=0; i < this.nPages; i++) {\n",
       "                this.pid = i;\n",
       "                for (var j=0; j < this.cids[i].length; j++) {\n",
       "                    this.cxid = j;\n",
       "                    for (var k=0; k < this.cids[i][j].length; k++) {\n",
       "                        this.cid = k;\n",
       "                        if (this.cids[i][j][k] in this.labels) {\n",
       "                            this.markCurrentCandidate(false);\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "            this.pid  = 0;\n",
       "            this.cxid = 0;\n",
       "            this.cid  = 0;\n",
       "\n",
       "            // Enable button functionality for navigation\n",
       "            var that = this;\n",
       "            this.$el.find(\"#next-cand\").click(function() {\n",
       "                that.switchCandidate(1);\n",
       "            });\n",
       "            this.$el.find(\"#prev-cand\").click(function() {\n",
       "                that.switchCandidate(-1);\n",
       "            });\n",
       "            this.$el.find(\"#next-context\").click(function() {\n",
       "                that.switchContext(1);\n",
       "            });\n",
       "            this.$el.find(\"#prev-context\").click(function() {\n",
       "                that.switchContext(-1);\n",
       "            });\n",
       "            this.$el.find(\"#next-page\").click(function() {\n",
       "                that.switchPage(1);\n",
       "            });\n",
       "            this.$el.find(\"#prev-page\").click(function() {\n",
       "                that.switchPage(-1);\n",
       "            });\n",
       "            this.$el.find(\"#label-true\").click(function() {\n",
       "                that.labelCandidate(true, true);\n",
       "            });\n",
       "            this.$el.find(\"#label-false\").click(function() {\n",
       "                that.labelCandidate(false, true);\n",
       "            });\n",
       "\n",
       "            // Arrow key functionality\n",
       "            this.$el.keydown(function(e) {\n",
       "                switch(e.which) {\n",
       "                    case 74: // j\n",
       "                    that.switchCandidate(-1);\n",
       "                    break;\n",
       "\n",
       "                    case 73: // i\n",
       "                    that.switchPage(-1);\n",
       "                    break;\n",
       "\n",
       "                    case 76: // l\n",
       "                    that.switchCandidate(1);\n",
       "                    break;\n",
       "\n",
       "                    case 75: // k\n",
       "                    that.switchPage(1);\n",
       "                    break;\n",
       "\n",
       "                    case 84: // t\n",
       "                    that.labelCandidate(true, true);\n",
       "                    break;\n",
       "\n",
       "                    case 70: // f\n",
       "                    that.labelCandidate(false, true);\n",
       "                    break;\n",
       "                }\n",
       "            });\n",
       "\n",
       "            // Show the first page and highlight the first candidate\n",
       "            this.$el.find(\"#viewer-page-0\").show();\n",
       "            this.switchCandidate(0);\n",
       "        },\n",
       "\n",
       "        // Get candidate selector for currently selected candidate, escaping id properly\n",
       "        getCandidate: function() {\n",
       "            return this.$el.find(\".\"+this.cids[this.pid][this.cxid][this.cid]);\n",
       "        },  \n",
       "\n",
       "        // Color the candidate correctly according to registered label, as well as set highlighting\n",
       "        markCurrentCandidate: function(highlight) {\n",
       "            var cid  = this.cids[this.pid][this.cxid][this.cid];\n",
       "            var tags = this.$el.find(\".\"+cid);\n",
       "\n",
       "            // Clear color classes\n",
       "            tags.removeClass(\"candidate-h\");\n",
       "            tags.removeClass(\"true-candidate\");\n",
       "            tags.removeClass(\"true-candidate-h\");\n",
       "            tags.removeClass(\"false-candidate\");\n",
       "            tags.removeClass(\"false-candidate-h\");\n",
       "            tags.removeClass(\"highlighted\");\n",
       "\n",
       "            if (highlight) {\n",
       "                if (cid in this.labels) {\n",
       "                    tags.addClass(String(this.labels[cid]) + \"-candidate-h\");\n",
       "                } else {\n",
       "                    tags.addClass(\"candidate-h\");\n",
       "                }\n",
       "            \n",
       "            // If un-highlighting, leave with first non-null coloring\n",
       "            } else {\n",
       "                var that = this;\n",
       "                tags.each(function() {\n",
       "                    var cids = $(this).attr('class').split(/\\s+/).map(function(item) {\n",
       "                        return parseInt(item);\n",
       "                    });\n",
       "                    cids.sort();\n",
       "                    for (var i in cids) {\n",
       "                        if (cids[i] in that.labels) {\n",
       "                            var label = that.labels[cids[i]];\n",
       "                            $(this).addClass(String(label) + \"-candidate\");\n",
       "                            $(this).removeClass(String(!label) + \"-candidate\");\n",
       "                            break;\n",
       "                        }\n",
       "                    }\n",
       "                });\n",
       "            }\n",
       "\n",
       "            // Extra highlighting css\n",
       "            if (highlight) {\n",
       "                tags.addClass(\"highlighted\");\n",
       "            }\n",
       "\n",
       "            // Classes for showing direction of relation\n",
       "            if (highlight) {\n",
       "                this.$el.find(\".\"+cid+\"-0\").addClass(\"left-candidate\");\n",
       "                this.$el.find(\".\"+cid+\"-1\").addClass(\"right-candidate\");\n",
       "            } else {\n",
       "                this.$el.find(\".\"+cid+\"-0\").removeClass(\"left-candidate\");\n",
       "                this.$el.find(\".\"+cid+\"-1\").removeClass(\"right-candidate\");\n",
       "            }\n",
       "        },\n",
       "\n",
       "        // Cycle through candidates and highlight, by increment inc\n",
       "        switchCandidate: function(inc) {\n",
       "            var N = this.cids[this.pid].length\n",
       "            var M = this.cids[this.pid][this.cxid].length;\n",
       "            if (N == 0 || M == 0) { return false; }\n",
       "\n",
       "            // Clear highlighting from previous candidate\n",
       "            if (inc != 0) {\n",
       "                this.markCurrentCandidate(false);\n",
       "\n",
       "                // Increment the cid counter\n",
       "\n",
       "                // Move to next context\n",
       "                if (this.cid + inc >= M) {\n",
       "                    while (this.cid + inc >= M) {\n",
       "                        \n",
       "                        // At last context on page, halt\n",
       "                        if (this.cxid == N - 1) {\n",
       "                            this.cid = M - 1;\n",
       "                            inc = 0;\n",
       "                            break;\n",
       "                        \n",
       "                        // Increment to next context\n",
       "                        } else {\n",
       "                            inc -= M - this.cid;\n",
       "                            this.cxid += 1;\n",
       "                            M = this.cids[this.pid][this.cxid].length;\n",
       "                            this.cid = 0;\n",
       "                        }\n",
       "                    }\n",
       "\n",
       "                // Move to previous context\n",
       "                } else if (this.cid + inc < 0) {\n",
       "                    while (this.cid + inc < 0) {\n",
       "                        \n",
       "                        // At first context on page, halt\n",
       "                        if (this.cxid == 0) {\n",
       "                            this.cid = 0;\n",
       "                            inc = 0;\n",
       "                            break;\n",
       "                        \n",
       "                        // Increment to previous context\n",
       "                        } else {\n",
       "                            inc += this.cid + 1;\n",
       "                            this.cxid -= 1;\n",
       "                            M = this.cids[this.pid][this.cxid].length;\n",
       "                            this.cid = M - 1;\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "\n",
       "                // Move within current context\n",
       "                this.cid += inc;\n",
       "            }\n",
       "            this.markCurrentCandidate(true);\n",
       "\n",
       "            // Push this new cid to the model\n",
       "            this.model.set('_selected_cid', this.cids[this.pid][this.cxid][this.cid]);\n",
       "            this.touch();\n",
       "        },\n",
       "\n",
       "        // Switch through contexts\n",
       "        switchContext: function(inc) {\n",
       "            this.markCurrentCandidate(false);\n",
       "\n",
       "            // Iterate context on this page\n",
       "            var M = this.cids[this.pid].length;\n",
       "            if (this.cxid + inc < 0) {\n",
       "                this.cxid = 0;\n",
       "            } else if (this.cxid + inc >= M) {\n",
       "                this.cxid = M - 1;\n",
       "            } else {\n",
       "                this.cxid += inc;\n",
       "            }\n",
       "\n",
       "            // Reset cid and set to first candidate\n",
       "            this.cid = 0;\n",
       "            this.switchCandidate(0);\n",
       "        },\n",
       "\n",
       "        // Switch through pages\n",
       "        switchPage: function(inc) {\n",
       "            this.markCurrentCandidate(false);\n",
       "            this.$el.find(\".viewer-page\").hide();\n",
       "            if (this.pid + inc < 0) {\n",
       "                this.pid = 0;\n",
       "            } else if (this.pid + inc > this.nPages - 1) {\n",
       "                this.pid = this.nPages - 1;\n",
       "            } else {\n",
       "                this.pid += inc;\n",
       "            }\n",
       "            this.$el.find(\"#viewer-page-\"+this.pid).show();\n",
       "\n",
       "            // Show pagination\n",
       "            this.$el.find(\"#page\").html(this.pid);\n",
       "\n",
       "            // Reset cid and set to first candidate\n",
       "            this.cid = 0;\n",
       "            this.cxid = 0;\n",
       "            this.switchCandidate(0);\n",
       "        },\n",
       "\n",
       "        // Label currently-selected candidate\n",
       "        labelCandidate: function(label, highlighted) {\n",
       "            var c    = this.getCandidate();\n",
       "            var cid  = this.cids[this.pid][this.cxid][this.cid];\n",
       "            var cl   = String(label) + \"-candidate\";\n",
       "            var clh  = String(label) + \"-candidate-h\";\n",
       "            var cln  = String(!label) + \"-candidate\";\n",
       "            var clnh = String(!label) + \"-candidate-h\";\n",
       "\n",
       "            // Toggle label highlighting\n",
       "            if (c.hasClass(cl) || c.hasClass(clh)) {\n",
       "                c.removeClass(cl);\n",
       "                c.removeClass(clh);\n",
       "                if (highlighted) {\n",
       "                    c.addClass(\"candidate-h\");\n",
       "                }\n",
       "                this.labels[cid] = null;\n",
       "                this.send({event: 'delete_label', cid: cid});\n",
       "            } else {\n",
       "                c.removeClass(cln);\n",
       "                c.removeClass(clnh);\n",
       "                if (highlighted) {\n",
       "                    c.addClass(clh);\n",
       "                } else {\n",
       "                    c.addClass(cl);\n",
       "                }\n",
       "                this.labels[cid] = label;\n",
       "                this.send({event: 'set_label', cid: cid, value: label});\n",
       "            }\n",
       "\n",
       "            // Set the label and pass back to the model\n",
       "            this.model.set('_labels_serialized', this.serializeDict(this.labels));\n",
       "            this.touch();\n",
       "        },\n",
       "\n",
       "        // Serialization of hash maps, because traitlets Dict doesn't seem to work...\n",
       "        serializeDict: function(d) {\n",
       "            var s = [];\n",
       "            for (var key in d) {\n",
       "                s.push(key+\"~~\"+d[key]);\n",
       "            }\n",
       "            return s.join();\n",
       "        },\n",
       "\n",
       "        // Deserialization of hash maps\n",
       "        deserializeDict: function(s) {\n",
       "            var d = {};\n",
       "            var entries = s.split(/,/);\n",
       "            var kv;\n",
       "            for (var i in entries) {\n",
       "                kv = entries[i].split(/~~/);\n",
       "                if (kv[1] == \"true\") {\n",
       "                    d[kv[0]] = true;\n",
       "                } else if (kv[1] == \"false\") {\n",
       "                    d[kv[0]] = false;\n",
       "                }\n",
       "            }\n",
       "            return d;\n",
       "        },\n",
       "    });\n",
       "\n",
       "    return {\n",
       "        ViewerView: ViewerView\n",
       "    };\n",
       "});\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "23a09a7d35494732ade182bc26edf15b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SentenceNgramViewer(cids=[[[767, 1579, 1970, 2250, 2385, 2445], [2020, 2021], [477]], [[847, 1668], [65, 66], …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from snorkel.viewer import SentenceNgramViewer\n",
    "\n",
    "SentenceNgramViewer(labeled, session)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part II: Applying Labeling Functions\n",
    "\n",
    "We run the LFs over all training candidates, producing a set of Labels (Virus and Host) and LabelKeys (the names of the LFs) in the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up the label annotator class\n",
    "from snorkel.annotations import LabelAnnotator\n",
    "labeler = LabelAnnotator(lfs=LFs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clearing existing...\n",
      "Running UDF...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████| 3823/3823 [00:18<00:00, 208.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 18.5 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<3823x17 sparse matrix of type '<class 'numpy.int32'>'\n",
       "\twith 5733 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.seed(1701)\n",
    "%time L_train = labeler.apply(split=0)\n",
    "L_train"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the returned matrix is a special subclass of the `scipy.sparse.csr_matrix` class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "VirusHost(Span(\"b'JEV'\", sentence=12925, chars=[172,174], words=[31,31]), Span(\"b'human'\", sentence=12925, chars=[44,48], words=[8,8]))"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the candidate names and positions of any candidate in the set\n",
    "L_train.get_candidate(session, 0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LabelKey (LF_detect)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the LabelKey (the name of the LF used to identify the candidate)\n",
    "L_train.get_key(session, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Viewing statistics about the resulting label matrix:\n",
    "\n",
    "* **Coverage** is the fraction of candidates that the labeling function emits a non-zero label for.\n",
    "* **Overlap** is the fraction candidates that the labeling function emits a non-zero label for and that another labeling function emits a non-zero label for.\n",
    "* **Conflict** is the fraction candidates that the labeling function emits a non-zero label for and that another labeling function emits a *conflicting* non-zero label for."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LF_detect</th>\n",
       "      <td>0</td>\n",
       "      <td>0.072979</td>\n",
       "      <td>0.072979</td>\n",
       "      <td>0.021711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_detect2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.007847</td>\n",
       "      <td>0.007847</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_infect</th>\n",
       "      <td>2</td>\n",
       "      <td>0.057546</td>\n",
       "      <td>0.057546</td>\n",
       "      <td>0.012032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_isolate</th>\n",
       "      <td>3</td>\n",
       "      <td>0.077688</td>\n",
       "      <td>0.077688</td>\n",
       "      <td>0.026419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_positive</th>\n",
       "      <td>4</td>\n",
       "      <td>0.187026</td>\n",
       "      <td>0.187026</td>\n",
       "      <td>0.051269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_positive2</th>\n",
       "      <td>5</td>\n",
       "      <td>0.046037</td>\n",
       "      <td>0.045776</td>\n",
       "      <td>0.002354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_misc</th>\n",
       "      <td>6</td>\n",
       "      <td>0.118755</td>\n",
       "      <td>0.118755</td>\n",
       "      <td>0.023019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_v_cause_h</th>\n",
       "      <td>7</td>\n",
       "      <td>0.005755</td>\n",
       "      <td>0.005493</td>\n",
       "      <td>0.000785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_v_h</th>\n",
       "      <td>8</td>\n",
       "      <td>0.291133</td>\n",
       "      <td>0.150405</td>\n",
       "      <td>0.005231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_h_v</th>\n",
       "      <td>9</td>\n",
       "      <td>0.182056</td>\n",
       "      <td>0.065917</td>\n",
       "      <td>0.000262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_other_verbs</th>\n",
       "      <td>10</td>\n",
       "      <td>0.047345</td>\n",
       "      <td>0.046299</td>\n",
       "      <td>0.015694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_uncertain</th>\n",
       "      <td>11</td>\n",
       "      <td>0.003400</td>\n",
       "      <td>0.003400</td>\n",
       "      <td>0.003400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_far_v_h</th>\n",
       "      <td>12</td>\n",
       "      <td>0.268899</td>\n",
       "      <td>0.059377</td>\n",
       "      <td>0.057285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_far_h_v</th>\n",
       "      <td>13</td>\n",
       "      <td>0.113523</td>\n",
       "      <td>0.013079</td>\n",
       "      <td>0.013079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_neg_h</th>\n",
       "      <td>14</td>\n",
       "      <td>0.015433</td>\n",
       "      <td>0.010986</td>\n",
       "      <td>0.008632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_distant_supervision</th>\n",
       "      <td>15</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_neg_assertions</th>\n",
       "      <td>16</td>\n",
       "      <td>0.003662</td>\n",
       "      <td>0.002616</td>\n",
       "      <td>0.002354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         j  Coverage  Overlaps  Conflicts\n",
       "LF_detect                0  0.072979  0.072979   0.021711\n",
       "LF_detect2               1  0.007847  0.007847   0.000000\n",
       "LF_infect                2  0.057546  0.057546   0.012032\n",
       "LF_isolate               3  0.077688  0.077688   0.026419\n",
       "LF_positive              4  0.187026  0.187026   0.051269\n",
       "LF_positive2             5  0.046037  0.045776   0.002354\n",
       "LF_misc                  6  0.118755  0.118755   0.023019\n",
       "LF_v_cause_h             7  0.005755  0.005493   0.000785\n",
       "LF_v_h                   8  0.291133  0.150405   0.005231\n",
       "LF_h_v                   9  0.182056  0.065917   0.000262\n",
       "LF_other_verbs          10  0.047345  0.046299   0.015694\n",
       "LF_uncertain            11  0.003400  0.003400   0.003400\n",
       "LF_far_v_h              12  0.268899  0.059377   0.057285\n",
       "LF_far_h_v              13  0.113523  0.013079   0.013079\n",
       "LF_neg_h                14  0.015433  0.010986   0.008632\n",
       "LF_distant_supervision  15  0.000523  0.000523   0.000000\n",
       "LF_neg_assertions       16  0.003662  0.002616   0.002354"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L_train.lf_stats(session)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part III: Checking Against Gold Labels (Hand Labeled Set)\n",
    "- Run the labeler on the development set\n",
    "- Load in some external labels:\n",
    "\n",
    "### Load Gold Labels\n",
    "Gold labels are a _small_ set of examples (here, a subset of our training set) which we label by hand and use to help us develop and refine labeling functions. Unlike the _test set_, which we do not look at and use for final evaluation, we can inspect the development set while writing labeling functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<430x1 sparse matrix of type '<class 'numpy.int32'>'\n",
       "\twith 126 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from snorkel.annotations import load_gold_labels\n",
    "L_gold_dev = load_gold_labels(session, annotator_name = \"gold\", split=1)\n",
    "L_gold_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clearing existing...\n",
      "Running UDF...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 430/430 [00:02<00:00, 163.07it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 2.71 s\n"
     ]
    }
   ],
   "source": [
    "%time L_dev = labeler.apply_existing(split=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>FN</th>\n",
       "      <th>TN</th>\n",
       "      <th>Empirical Acc.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LF_detect</th>\n",
       "      <td>0</td>\n",
       "      <td>0.032558</td>\n",
       "      <td>0.032558</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_detect2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.006977</td>\n",
       "      <td>0.006977</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_infect</th>\n",
       "      <td>2</td>\n",
       "      <td>0.025581</td>\n",
       "      <td>0.025581</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_isolate</th>\n",
       "      <td>3</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_positive</th>\n",
       "      <td>4</td>\n",
       "      <td>0.060465</td>\n",
       "      <td>0.060465</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.764706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_positive2</th>\n",
       "      <td>5</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_misc</th>\n",
       "      <td>6</td>\n",
       "      <td>0.148837</td>\n",
       "      <td>0.148837</td>\n",
       "      <td>0.055814</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.937500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_v_cause_h</th>\n",
       "      <td>7</td>\n",
       "      <td>0.004651</td>\n",
       "      <td>0.004651</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_v_h</th>\n",
       "      <td>8</td>\n",
       "      <td>0.286047</td>\n",
       "      <td>0.086047</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>31</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.738095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_h_v</th>\n",
       "      <td>9</td>\n",
       "      <td>0.262791</td>\n",
       "      <td>0.083721</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.818182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_other_verbs</th>\n",
       "      <td>10</td>\n",
       "      <td>0.039535</td>\n",
       "      <td>0.034884</td>\n",
       "      <td>0.004651</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_uncertain</th>\n",
       "      <td>11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_far_v_h</th>\n",
       "      <td>12</td>\n",
       "      <td>0.155814</td>\n",
       "      <td>0.027907</td>\n",
       "      <td>0.027907</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>0.904762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_far_h_v</th>\n",
       "      <td>13</td>\n",
       "      <td>0.218605</td>\n",
       "      <td>0.030233</td>\n",
       "      <td>0.030233</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_neg_h</th>\n",
       "      <td>14</td>\n",
       "      <td>0.004651</td>\n",
       "      <td>0.004651</td>\n",
       "      <td>0.004651</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_distant_supervision</th>\n",
       "      <td>15</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_neg_assertions</th>\n",
       "      <td>16</td>\n",
       "      <td>0.002326</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         j  Coverage  Overlaps  Conflicts  TP  FP  FN  TN  \\\n",
       "LF_detect                0  0.032558  0.032558   0.000000   6   3   0   0   \n",
       "LF_detect2               1  0.006977  0.006977   0.000000   2   1   0   0   \n",
       "LF_infect                2  0.025581  0.025581   0.000000   7   0   0   0   \n",
       "LF_isolate               3  0.002326  0.002326   0.000000   0   1   0   0   \n",
       "LF_positive              4  0.060465  0.060465   0.000000  13   4   0   0   \n",
       "LF_positive2             5  0.023256  0.023256   0.000000   4   1   0   0   \n",
       "LF_misc                  6  0.148837  0.148837   0.055814  15   1   0   0   \n",
       "LF_v_cause_h             7  0.004651  0.004651   0.000000   2   0   0   0   \n",
       "LF_v_h                   8  0.286047  0.086047   0.000000  31  11   0   0   \n",
       "LF_h_v                   9  0.262791  0.083721   0.002326  36   8   0   0   \n",
       "LF_other_verbs          10  0.039535  0.034884   0.004651   5   2   0   0   \n",
       "LF_uncertain            11  0.000000  0.000000   0.000000   0   0   0   0   \n",
       "LF_far_v_h              12  0.155814  0.027907   0.027907   0   0   2  19   \n",
       "LF_far_h_v              13  0.218605  0.030233   0.030233   0   0   1   5   \n",
       "LF_neg_h                14  0.004651  0.004651   0.004651   0   0   0   0   \n",
       "LF_distant_supervision  15  0.002326  0.002326   0.000000   0   0   0   0   \n",
       "LF_neg_assertions       16  0.002326  0.000000   0.000000   0   0   0   0   \n",
       "\n",
       "                        Empirical Acc.  \n",
       "LF_detect                     0.666667  \n",
       "LF_detect2                    0.666667  \n",
       "LF_infect                     1.000000  \n",
       "LF_isolate                    0.000000  \n",
       "LF_positive                   0.764706  \n",
       "LF_positive2                  0.800000  \n",
       "LF_misc                       0.937500  \n",
       "LF_v_cause_h                  1.000000  \n",
       "LF_v_h                        0.738095  \n",
       "LF_h_v                        0.818182  \n",
       "LF_other_verbs                0.714286  \n",
       "LF_uncertain                       NaN  \n",
       "LF_far_v_h                    0.904762  \n",
       "LF_far_h_v                    0.833333  \n",
       "LF_neg_h                           NaN  \n",
       "LF_distant_supervision             NaN  \n",
       "LF_neg_assertions                  NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Label Matrix Empirical Accuracies\n",
    "\n",
    "L_dev.lf_stats(session, labels=L_gold_dev.toarray().ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Iterating on Labeling Function Design:\n",
    "When writing labeling functions, you will want to iterate on the process outlined above several times. You should focus on tuning individual LFs, based on emprical accuracy metrics, and adding new LFs to improve coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "### See Notebook Part 3 for Generative Model Training"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (snorkel-env)",
   "language": "python",
   "name": "snorkel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
